{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "automated-uganda",
   "metadata": {},
   "source": [
    "# Notebook dev"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unlimited-visibility",
   "metadata": {},
   "source": [
    "NB : all variables & filters using the word CDS are corresponding to the TER concept in the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "whole-error",
   "metadata": {},
   "outputs": [],
   "source": [
    "# General imports\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import scipy\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "sys.path.append('../')\n",
    "# Other imports\n",
    "\n",
    "tqdm.pandas()\n",
    "import yaml\n",
    "import json\n",
    "\n",
    "\n",
    "## YAML FILES CONFIG\n",
    "yaml = yaml.load(open(\"config/config_files.yaml\"), Loader=yaml.FullLoader)\n",
    "base_dir = yaml['base_directory']\n",
    "sys.exit('EXIT : Need to specify the base_directory in config file : \"conf_files.yaml\"') if base_dir == 'TO_CHANGE' else None\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "descending-patient",
   "metadata": {},
   "source": [
    "# Define Miso & Siso"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "laden-jungle",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/weber/.conda/envs/ExoCarto/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes : 19664\n",
      "mRNA : 61295\n",
      "Genes :  Raw_Miso_siso\n",
      "Miso    13552\n",
      "Siso     6112\n",
      "Name: GeneID, dtype: int64\n",
      "mRNA :  Raw_Miso_siso\n",
      "Miso    55146\n",
      "Siso     6112\n",
      "Name: transcript_count_raw, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "genes_raw = pd.read_csv(base_dir + yaml['External']['biomart_ensembl'], compression='gzip', sep='\\t')\n",
    "genes_raw['Chromosome/scaffold name'] = genes_raw['Chromosome/scaffold name'].astype(str)\n",
    "genes_raw = genes_raw.loc[~genes_raw['Chromosome/scaffold name'].str.contains('CHR|\\.')].copy()\n",
    "genes = genes_raw[['Gene stable ID', 'Gene name', 'Chromosome/scaffold name', 'Gene start (bp)', 'Gene end (bp)']].drop_duplicates().sort_values(by='Gene name')\n",
    "genes['Gene_length'] = genes['Gene end (bp)'] - genes['Gene start (bp)']\n",
    "print('Genes : {}'.format(genes.shape[0]), )\n",
    "\n",
    "mrna = genes_raw[['Gene stable ID', 'Transcript stable ID', 'Protein stable ID', 'Transcript start (bp)', 'Transcript end (bp)', 'Transcription start site (TSS)', 'Transcript length (including UTRs and CDS)', 'Transcript support level (TSL)', 'GENCODE basic annotation', 'APPRIS annotation']]\n",
    "print('mRNA : {}'.format(mrna.shape[0]), )\n",
    "\n",
    "miso_siso_raw = mrna[['Gene stable ID', 'Transcript stable ID']].groupby('Gene stable ID')['Transcript stable ID'].nunique().reset_index()\n",
    "miso_siso_raw.loc[miso_siso_raw['Transcript stable ID'] > 1, 'Raw_Miso_siso'] = 'Miso'\n",
    "miso_siso_raw.loc[miso_siso_raw['Transcript stable ID'] == 1, 'Raw_Miso_siso'] = 'Siso'\n",
    "miso_siso_raw = miso_siso_raw.rename({'Gene stable ID': 'GeneID', 'Transcript stable ID' : 'transcript_count_raw'}, axis=1)\n",
    "# print('MISOG SISOG : {}'.format(mrna.shape[0]), mrna.columns)\n",
    "print('Genes : ', miso_siso_raw.groupby('Raw_Miso_siso')['GeneID'].nunique())\n",
    "print('mRNA : ',miso_siso_raw.groupby('Raw_Miso_siso')['transcript_count_raw'].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "fifty-alfred",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Transcript start (bp)</th>\n",
       "      <th>Transcript end (bp)</th>\n",
       "      <th>...</th>\n",
       "      <th>Source of gene name</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>Source of transcript name</th>\n",
       "      <th>Gene % GC content</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Source (gene)</th>\n",
       "      <th>Source (transcript)</th>\n",
       "      <th>CCDS ID</th>\n",
       "      <th>UniProtKB isoform ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>1</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-ND1-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>47.70</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>insdc</td>\n",
       "      <td>insdc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENST00000361453</td>\n",
       "      <td>ENSP00000355046</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>MT</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>1</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-ND2-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>42.99</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>insdc</td>\n",
       "      <td>insdc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>ENST00000361624</td>\n",
       "      <td>ENSP00000354499</td>\n",
       "      <td>mitochondrially encoded cytochrome c oxidase I...</td>\n",
       "      <td>MT</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>1</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-CO1-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>46.24</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>insdc</td>\n",
       "      <td>insdc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198712</td>\n",
       "      <td>ENST00000361739</td>\n",
       "      <td>ENSP00000354876</td>\n",
       "      <td>mitochondrially encoded cytochrome c oxidase I...</td>\n",
       "      <td>MT</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>1</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-CO2-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>46.20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>insdc</td>\n",
       "      <td>insdc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>ENST00000361851</td>\n",
       "      <td>ENSP00000355265</td>\n",
       "      <td>mitochondrially encoded ATP synthase membrane ...</td>\n",
       "      <td>MT</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>1</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>MT-ATP8-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>39.61</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>insdc</td>\n",
       "      <td>insdc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68279</th>\n",
       "      <td>ENSG00000116754</td>\n",
       "      <td>ENST00000395136</td>\n",
       "      <td>ENSP00000378568</td>\n",
       "      <td>serine and arginine rich splicing factor 11 [S...</td>\n",
       "      <td>1</td>\n",
       "      <td>70205682</td>\n",
       "      <td>70253052</td>\n",
       "      <td>1</td>\n",
       "      <td>70221396</td>\n",
       "      <td>70250805</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>SRSF11-204</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>37.01</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68280</th>\n",
       "      <td>ENSG00000283324</td>\n",
       "      <td>ENST00000636087</td>\n",
       "      <td>ENSP00000490418</td>\n",
       "      <td>cortexin domain containing 2 [Source:HGNC Symb...</td>\n",
       "      <td>1</td>\n",
       "      <td>150887136</td>\n",
       "      <td>150913292</td>\n",
       "      <td>1</td>\n",
       "      <td>150887136</td>\n",
       "      <td>150913292</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>CTXND2-201</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>42.36</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>havana</td>\n",
       "      <td>havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68281</th>\n",
       "      <td>ENSG00000077585</td>\n",
       "      <td>ENST00000366592</td>\n",
       "      <td>ENSP00000355551</td>\n",
       "      <td>G protein-coupled receptor 137B [Source:HGNC S...</td>\n",
       "      <td>1</td>\n",
       "      <td>236142505</td>\n",
       "      <td>236221865</td>\n",
       "      <td>1</td>\n",
       "      <td>236142539</td>\n",
       "      <td>236208907</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>GPR137B-202</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>43.63</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CCDS1609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68282</th>\n",
       "      <td>ENSG00000086619</td>\n",
       "      <td>ENST00000354619</td>\n",
       "      <td>ENSP00000346635</td>\n",
       "      <td>endoplasmic reticulum oxidoreductase 1 beta [S...</td>\n",
       "      <td>1</td>\n",
       "      <td>236214681</td>\n",
       "      <td>236282019</td>\n",
       "      <td>-1</td>\n",
       "      <td>236215101</td>\n",
       "      <td>236281958</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ERO1B-202</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>37.46</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CCDS31064</td>\n",
       "      <td>Q86YB8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68283</th>\n",
       "      <td>ENSG00000086619</td>\n",
       "      <td>ENST00000687487</td>\n",
       "      <td>ENSP00000510551</td>\n",
       "      <td>endoplasmic reticulum oxidoreductase 1 beta [S...</td>\n",
       "      <td>1</td>\n",
       "      <td>236214681</td>\n",
       "      <td>236282019</td>\n",
       "      <td>-1</td>\n",
       "      <td>236218244</td>\n",
       "      <td>236275679</td>\n",
       "      <td>...</td>\n",
       "      <td>HGNC Symbol</td>\n",
       "      <td>ERO1B-206</td>\n",
       "      <td>Transcript name</td>\n",
       "      <td>37.46</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>havana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61295 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Transcript stable ID Protein stable ID  \\\n",
       "0      ENSG00000198888      ENST00000361390   ENSP00000354687   \n",
       "1      ENSG00000198763      ENST00000361453   ENSP00000355046   \n",
       "2      ENSG00000198804      ENST00000361624   ENSP00000354499   \n",
       "3      ENSG00000198712      ENST00000361739   ENSP00000354876   \n",
       "4      ENSG00000228253      ENST00000361851   ENSP00000355265   \n",
       "...                ...                  ...               ...   \n",
       "68279  ENSG00000116754      ENST00000395136   ENSP00000378568   \n",
       "68280  ENSG00000283324      ENST00000636087   ENSP00000490418   \n",
       "68281  ENSG00000077585      ENST00000366592   ENSP00000355551   \n",
       "68282  ENSG00000086619      ENST00000354619   ENSP00000346635   \n",
       "68283  ENSG00000086619      ENST00000687487   ENSP00000510551   \n",
       "\n",
       "                                        Gene description  \\\n",
       "0      mitochondrially encoded NADH:ubiquinone oxidor...   \n",
       "1      mitochondrially encoded NADH:ubiquinone oxidor...   \n",
       "2      mitochondrially encoded cytochrome c oxidase I...   \n",
       "3      mitochondrially encoded cytochrome c oxidase I...   \n",
       "4      mitochondrially encoded ATP synthase membrane ...   \n",
       "...                                                  ...   \n",
       "68279  serine and arginine rich splicing factor 11 [S...   \n",
       "68280  cortexin domain containing 2 [Source:HGNC Symb...   \n",
       "68281  G protein-coupled receptor 137B [Source:HGNC S...   \n",
       "68282  endoplasmic reticulum oxidoreductase 1 beta [S...   \n",
       "68283  endoplasmic reticulum oxidoreductase 1 beta [S...   \n",
       "\n",
       "      Chromosome/scaffold name  Gene start (bp)  Gene end (bp)  Strand  \\\n",
       "0                           MT             3307           4262       1   \n",
       "1                           MT             4470           5511       1   \n",
       "2                           MT             5904           7445       1   \n",
       "3                           MT             7586           8269       1   \n",
       "4                           MT             8366           8572       1   \n",
       "...                        ...              ...            ...     ...   \n",
       "68279                        1         70205682       70253052       1   \n",
       "68280                        1        150887136      150913292       1   \n",
       "68281                        1        236142505      236221865       1   \n",
       "68282                        1        236214681      236282019      -1   \n",
       "68283                        1        236214681      236282019      -1   \n",
       "\n",
       "       Transcript start (bp)  Transcript end (bp)  ...  Source of gene name  \\\n",
       "0                       3307                 4262  ...          HGNC Symbol   \n",
       "1                       4470                 5511  ...          HGNC Symbol   \n",
       "2                       5904                 7445  ...          HGNC Symbol   \n",
       "3                       7586                 8269  ...          HGNC Symbol   \n",
       "4                       8366                 8572  ...          HGNC Symbol   \n",
       "...                      ...                  ...  ...                  ...   \n",
       "68279               70221396             70250805  ...          HGNC Symbol   \n",
       "68280              150887136            150913292  ...          HGNC Symbol   \n",
       "68281              236142539            236208907  ...          HGNC Symbol   \n",
       "68282              236215101            236281958  ...          HGNC Symbol   \n",
       "68283              236218244            236275679  ...          HGNC Symbol   \n",
       "\n",
       "       Transcript name Source of transcript name Gene % GC content  \\\n",
       "0           MT-ND1-201           Transcript name             47.70   \n",
       "1           MT-ND2-201           Transcript name             42.99   \n",
       "2           MT-CO1-201           Transcript name             46.24   \n",
       "3           MT-CO2-201           Transcript name             46.20   \n",
       "4          MT-ATP8-201           Transcript name             39.61   \n",
       "...                ...                       ...               ...   \n",
       "68279       SRSF11-204           Transcript name             37.01   \n",
       "68280       CTXND2-201           Transcript name             42.36   \n",
       "68281      GPR137B-202           Transcript name             43.63   \n",
       "68282        ERO1B-202           Transcript name             37.46   \n",
       "68283        ERO1B-206           Transcript name             37.46   \n",
       "\n",
       "            Gene type  Transcript type   Source (gene) Source (transcript)  \\\n",
       "0      protein_coding   protein_coding           insdc               insdc   \n",
       "1      protein_coding   protein_coding           insdc               insdc   \n",
       "2      protein_coding   protein_coding           insdc               insdc   \n",
       "3      protein_coding   protein_coding           insdc               insdc   \n",
       "4      protein_coding   protein_coding           insdc               insdc   \n",
       "...               ...              ...             ...                 ...   \n",
       "68279  protein_coding   protein_coding  ensembl_havana              havana   \n",
       "68280  protein_coding   protein_coding          havana              havana   \n",
       "68281  protein_coding   protein_coding  ensembl_havana      ensembl_havana   \n",
       "68282  protein_coding   protein_coding  ensembl_havana      ensembl_havana   \n",
       "68283  protein_coding   protein_coding  ensembl_havana              havana   \n",
       "\n",
       "         CCDS ID UniProtKB isoform ID  \n",
       "0            NaN                  NaN  \n",
       "1            NaN                  NaN  \n",
       "2            NaN                  NaN  \n",
       "3            NaN                  NaN  \n",
       "4            NaN                  NaN  \n",
       "...          ...                  ...  \n",
       "68279        NaN                  NaN  \n",
       "68280        NaN                  NaN  \n",
       "68281   CCDS1609                  NaN  \n",
       "68282  CCDS31064             Q86YB8-1  \n",
       "68283        NaN                  NaN  \n",
       "\n",
       "[61295 rows x 29 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "attended-apollo",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTEx Genes :  16172\n",
      "GTEx Genes :  Miso_siso\n",
      "Miso    9826\n",
      "Siso    6346\n",
      "Name: GeneID, dtype: int64\n",
      "GTEx mRNA :  37433\n",
      "GTEx mRNA :  Miso_siso\n",
      "Miso    31087\n",
      "Siso     6346\n",
      "Name: transcript_count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "gtex_checked = pd.read_csv('/gstock/EXOTIC/data/EXPRESSION/GTEx_V8_transcript_checking.gct.gz', compression='gzip', sep='\\t')\n",
    "gtex_checked['transcript_id'] = gtex_checked['transcript_id'].apply(lambda r : r.split('.')[0])\n",
    "\n",
    "mrna = mrna.rename({'Gene stable ID': 'GeneID', 'Transcript stable ID' : 'transcript_id'}, axis=1)\n",
    "\n",
    "mrna = mrna.loc[mrna['transcript_id'].isin(gtex_checked.loc[gtex_checked['check'] == True]['transcript_id'].values.tolist())]\n",
    "\n",
    "\n",
    "miso_siso = mrna[['GeneID', 'transcript_id']].groupby('GeneID')['transcript_id'].nunique().reset_index()\n",
    "miso_siso.loc[miso_siso['transcript_id'] > 1, 'Miso_siso'] = 'Miso'\n",
    "miso_siso.loc[miso_siso['transcript_id'] == 1, 'Miso_siso'] = 'Siso'\n",
    "miso_siso = miso_siso.rename({'transcript_id' : 'transcript_count'}, axis=1)\n",
    "miso_siso.groupby('Miso_siso')['transcript_count'].sum()\n",
    "print('GTEx Genes : ', miso_siso.groupby('Miso_siso')['GeneID'].nunique().sum())\n",
    "print('GTEx Genes : ', miso_siso.groupby('Miso_siso')['GeneID'].nunique())\n",
    "print('GTEx mRNA : ',miso_siso.groupby('Miso_siso')['transcript_count'].sum().sum())\n",
    "print('GTEx mRNA : ',miso_siso.groupby('Miso_siso')['transcript_count'].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "atmospheric-escape",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2281\n"
     ]
    }
   ],
   "source": [
    "merge_raw_gtex = pd.merge(miso_siso, miso_siso_raw, on=['GeneID'])\n",
    "merge_raw_gtex.loc[merge_raw_gtex['Miso_siso'] == merge_raw_gtex['Raw_Miso_siso'], 'Same'] = True\n",
    "merge_raw_gtex.loc[merge_raw_gtex['Miso_siso'] != merge_raw_gtex['Raw_Miso_siso'], 'Same'] = False\n",
    "genes_to_keep = merge_raw_gtex.loc[merge_raw_gtex['Same'] == True]['GeneID'].values.tolist()\n",
    "genes_to_not_keep = merge_raw_gtex.loc[merge_raw_gtex['Same'] == False]['GeneID'].values.tolist()\n",
    "print(len(genes_to_not_keep))\n",
    "mrna = mrna.loc[mrna['GeneID'].isin(genes_to_keep)]\n",
    "mrna_miso_siso = pd.merge(mrna, miso_siso, on=['GeneID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "latin-syndication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso_siso\n",
       "Miso    31087\n",
       "Siso     4065\n",
       "Name: transcript_count, dtype: int64"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_miso_siso[['GeneID', 'transcript_count', 'Miso_siso']].drop_duplicates().groupby('Miso_siso')['transcript_count'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "decimal-editor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2281"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes_to_not_keep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ordinary-dairy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso    12107\n",
       "Siso     4065\n",
       "Name: Raw_Miso_siso, dtype: int64"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_raw_gtex.Raw_Miso_siso.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "forward-constant",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso    9826\n",
       "Siso    6346\n",
       "Name: Miso_siso, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_raw_gtex.Miso_siso.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "given-sudan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso    9826\n",
       "Siso    4065\n",
       "Name: Miso_siso, dtype: int64"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_raw_gtex.loc[merge_raw_gtex['Same'] == True]['Miso_siso'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "mineral-mainland",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35152"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_raw_gtex[['GeneID', 'transcript_count', 'Miso_siso']].drop_duplicates()['transcript_count'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "alternative-inclusion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso_siso\n",
       "Miso    31087\n",
       "Siso     4065\n",
       "Name: transcript_count, dtype: int64"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_raw_gtex.loc[merge_raw_gtex['Same'] == True][['GeneID', 'transcript_count', 'Miso_siso']].drop_duplicates().groupby('Miso_siso')['transcript_count'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "prepared-universal",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Miso_siso</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Miso</th>\n",
       "      <td>31111.0</td>\n",
       "      <td>3.975025</td>\n",
       "      <td>2.305804</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siso</th>\n",
       "      <td>4065.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             count      mean       std  min  25%  50%  75%   max\n",
       "Miso_siso                                                       \n",
       "Miso       31111.0  3.975025  2.305804  2.0  2.0  3.0  5.0  21.0\n",
       "Siso        4065.0  1.000000  0.000000  1.0  1.0  1.0  1.0   1.0"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_miso_siso.groupby('Miso_siso')['transcript_count'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "logical-actor",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "prerequisite-phenomenon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exons = pd.read_csv('/gstock/GeneIso/data/External/ENSEMBL/mart_export_exon_38.txt.gz', compression='gzip', sep='\\t')\n",
    "exons = pd.read_csv(base_dir + yaml['External']['biomart_ensembl_exons'], compression='gzip', sep='\\t')\n",
    "\n",
    "exons = exons.rename({'Gene stable ID': 'GeneID', 'Transcript stable ID' : 'transcript_id'}, axis=1)\n",
    "\n",
    "merge_exon_miso_siso = pd.merge(mrna_miso_siso, exons, on=['GeneID', 'transcript_id'])\n",
    "\n",
    "\n",
    "tmp = merge_exon_miso_siso[['GeneID', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'Miso_siso']].drop_duplicates().groupby(['Miso_siso', 'GeneID'])['Exon region start (bp)'].count().reset_index()\n",
    "sisog_single_ter = tmp.loc[tmp['Exon region start (bp)'] == 1]['GeneID'].values.tolist()\n",
    "print(len(sisog_single_ter))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "declared-drill",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gene stable ID',\n",
       " 'Transcript stable ID',\n",
       " 'Exon region start (bp)',\n",
       " 'Exon region end (bp)',\n",
       " \"5' UTR start\",\n",
       " \"5' UTR end\",\n",
       " \"3' UTR start\",\n",
       " \"3' UTR end\",\n",
       " 'CDS Length',\n",
       " 'Strand',\n",
       " 'Transcription start site (TSS)',\n",
       " 'Transcript length (including UTRs and CDS)',\n",
       " 'Transcript start (bp)',\n",
       " 'Transcript end (bp)',\n",
       " 'Gene start (bp)',\n",
       " 'Gene end (bp)',\n",
       " 'Chromosome/scaffold name',\n",
       " 'CDS start',\n",
       " 'CDS end',\n",
       " 'Constitutive exon',\n",
       " 'Exon rank in transcript',\n",
       " 'Start phase',\n",
       " 'End phase',\n",
       " 'cDNA coding start',\n",
       " 'cDNA coding end',\n",
       " 'Genomic coding start',\n",
       " 'Genomic coding end',\n",
       " 'Exon stable ID']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "armed-daughter",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes :  13501\n"
     ]
    }
   ],
   "source": [
    "miso_siso_to_keep = miso_siso.loc[~miso_siso['GeneID'].isin(sisog_single_ter + genes_to_not_keep)]\n",
    "print('Genes : ', miso_siso_to_keep.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "compressed-occurrence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Miso_siso</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Miso</th>\n",
       "      <td>9826.0</td>\n",
       "      <td>82043.624364</td>\n",
       "      <td>143420.177937</td>\n",
       "      <td>704.0</td>\n",
       "      <td>15084.5</td>\n",
       "      <td>36573.5</td>\n",
       "      <td>89611.75</td>\n",
       "      <td>2473538.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siso</th>\n",
       "      <td>3675.0</td>\n",
       "      <td>51655.191837</td>\n",
       "      <td>90222.618571</td>\n",
       "      <td>489.0</td>\n",
       "      <td>9536.5</td>\n",
       "      <td>23037.0</td>\n",
       "      <td>56162.00</td>\n",
       "      <td>1899593.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            count          mean            std    min      25%      50%  \\\n",
       "Miso_siso                                                                 \n",
       "Miso       9826.0  82043.624364  143420.177937  704.0  15084.5  36573.5   \n",
       "Siso       3675.0  51655.191837   90222.618571  489.0   9536.5  23037.0   \n",
       "\n",
       "                75%        max  \n",
       "Miso_siso                       \n",
       "Miso       89611.75  2473538.0  \n",
       "Siso       56162.00  1899593.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_miso_siso = pd.merge(miso_siso_to_keep, genes.rename({'Gene stable ID': 'GeneID'}, axis=1), on ='GeneID')\n",
    "genes_miso_siso.groupby('Miso_siso')['Gene_length'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "according-latino",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Miso_siso</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Miso</th>\n",
       "      <td>31111.0</td>\n",
       "      <td>3069.642249</td>\n",
       "      <td>2360.576207</td>\n",
       "      <td>267.0</td>\n",
       "      <td>1537.0</td>\n",
       "      <td>2424.0</td>\n",
       "      <td>3925.0</td>\n",
       "      <td>34626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siso</th>\n",
       "      <td>3675.0</td>\n",
       "      <td>3728.145034</td>\n",
       "      <td>2771.967276</td>\n",
       "      <td>303.0</td>\n",
       "      <td>1805.0</td>\n",
       "      <td>2979.0</td>\n",
       "      <td>4865.0</td>\n",
       "      <td>30609.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             count         mean          std    min     25%     50%     75%  \\\n",
       "Miso_siso                                                                     \n",
       "Miso       31111.0  3069.642249  2360.576207  267.0  1537.0  2424.0  3925.0   \n",
       "Siso        3675.0  3728.145034  2771.967276  303.0  1805.0  2979.0  4865.0   \n",
       "\n",
       "               max  \n",
       "Miso_siso           \n",
       "Miso       34626.0  \n",
       "Siso       30609.0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_miso_siso = pd.merge(miso_siso_to_keep, mrna.rename({'Gene stable ID': 'GeneID'}, axis=1), on ='GeneID')\n",
    "mrna_miso_siso.groupby('Miso_siso')['Transcript length (including UTRs and CDS)'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "expensive-priest",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>transcript_count</th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>3</td>\n",
       "      <td>Miso</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>X</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>12883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>1</td>\n",
       "      <td>Siso</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>X</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "      <td>14949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>3</td>\n",
       "      <td>Miso</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>20</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50959140</td>\n",
       "      <td>24273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>3</td>\n",
       "      <td>Miso</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "      <td>44636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>4</td>\n",
       "      <td>Miso</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>1</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "      <td>192073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13496</th>\n",
       "      <td>ENSG00000286143</td>\n",
       "      <td>1</td>\n",
       "      <td>Siso</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>219491867</td>\n",
       "      <td>219498244</td>\n",
       "      <td>6377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13497</th>\n",
       "      <td>ENSG00000286185</td>\n",
       "      <td>1</td>\n",
       "      <td>Siso</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>149390623</td>\n",
       "      <td>149556361</td>\n",
       "      <td>165738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13498</th>\n",
       "      <td>ENSG00000286190</td>\n",
       "      <td>1</td>\n",
       "      <td>Siso</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>5499427</td>\n",
       "      <td>5501006</td>\n",
       "      <td>1579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13499</th>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>1</td>\n",
       "      <td>Siso</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>88523810</td>\n",
       "      <td>88708450</td>\n",
       "      <td>184640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13500</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>1</td>\n",
       "      <td>Siso</td>\n",
       "      <td>RPSAP58</td>\n",
       "      <td>19</td>\n",
       "      <td>23762944</td>\n",
       "      <td>23871162</td>\n",
       "      <td>108218</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13501 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID  transcript_count Miso_siso Gene name  \\\n",
       "0      ENSG00000000003                 3      Miso    TSPAN6   \n",
       "1      ENSG00000000005                 1      Siso      TNMD   \n",
       "2      ENSG00000000419                 3      Miso      DPM1   \n",
       "3      ENSG00000000457                 3      Miso     SCYL3   \n",
       "4      ENSG00000000460                 4      Miso  C1orf112   \n",
       "...                ...               ...       ...       ...   \n",
       "13496  ENSG00000286143                 1      Siso       NaN   \n",
       "13497  ENSG00000286185                 1      Siso       NaN   \n",
       "13498  ENSG00000286190                 1      Siso       NaN   \n",
       "13499  ENSG00000287542                 1      Siso       NaN   \n",
       "13500  ENSG00000288920                 1      Siso   RPSAP58   \n",
       "\n",
       "      Chromosome/scaffold name  Gene start (bp)  Gene end (bp)  Gene_length  \n",
       "0                            X        100627108      100639991        12883  \n",
       "1                            X        100584936      100599885        14949  \n",
       "2                           20         50934867       50959140        24273  \n",
       "3                            1        169849631      169894267        44636  \n",
       "4                            1        169662007      169854080       192073  \n",
       "...                        ...              ...            ...          ...  \n",
       "13496                        2        219491867      219498244         6377  \n",
       "13497                        1        149390623      149556361       165738  \n",
       "13498                       17          5499427        5501006         1579  \n",
       "13499                        4         88523810       88708450       184640  \n",
       "13500                       19         23762944       23871162       108218  \n",
       "\n",
       "[13501 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_miso_siso"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "piano-mailman",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_miso_siso.to_parquet('/gstock/GeneIso/V2/Genes.parquet')\n",
    "mrna_miso_siso.to_parquet('/gstock/GeneIso/V2/mRNA.parquet')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "operating-depression",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    9826.000000\n",
       "mean        3.163749\n",
       "std         1.602909\n",
       "min         2.000000\n",
       "25%         2.000000\n",
       "50%         3.000000\n",
       "75%         4.000000\n",
       "max        21.000000\n",
       "Name: transcript_id, dtype: float64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_miso_siso.loc[mrna_miso_siso['Miso_siso'] == 'Miso'].groupby('GeneID')['transcript_id'].nunique().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "dominant-export",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count   9,149.00\n",
       "mean        3.00\n",
       "std         1.00\n",
       "min         2.00\n",
       "25%         2.00\n",
       "50%         3.00\n",
       "75%         4.00\n",
       "max        21.00\n",
       "Name: transcript_count, dtype: float64"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miso_siso.loc[miso_siso['Miso_siso'] == 'Miso']['transcript_count'].describe().round()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "infectious-disposition",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_miso_siso_to_keep = pd.merge(exons, miso_siso_to_keep, on='GeneID')\n",
    "exon_miso_siso_to_keep['Exon_length'] = exon_miso_siso_to_keep['Exon region end (bp)'] - exon_miso_siso_to_keep['Exon region start (bp)']\n",
    "# exon_miso_siso_to_keep.drop_duplicates(subset=['GeneID', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)']).groupby(['Miso_siso'])['Exon region start (bp)'].count()\n",
    "# exon_miso_siso_to_keep.drop_duplicates(subset=['GeneID', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)']).groupby(['Miso_siso', 'GeneID'])['Exon region start (bp)'].count().groupby('Miso_siso').describe()\n",
    "# exon_miso_siso_to_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "quick-vinyl",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Miso_siso</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Miso</th>\n",
       "      <td>183882.0</td>\n",
       "      <td>409.412607</td>\n",
       "      <td>921.972520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>269.0</td>\n",
       "      <td>33289.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siso</th>\n",
       "      <td>36173.0</td>\n",
       "      <td>377.761314</td>\n",
       "      <td>961.647594</td>\n",
       "      <td>2.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>27302.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              count        mean         std  min   25%    50%    75%      max\n",
       "Miso_siso                                                                    \n",
       "Miso       183882.0  409.412607  921.972520  0.0  95.0  145.0  269.0  33289.0\n",
       "Siso        36173.0  377.761314  961.647594  2.0  97.0  137.0  203.0  27302.0"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_miso_siso_to_keep.drop_duplicates(subset=['GeneID', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)']).groupby('Miso_siso')['Exon_length'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "executed-ancient",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GeneID', 'transcript_id', 'Exon region start (bp)',\n",
       "       'Exon region end (bp)', '5' UTR start', '5' UTR end', '3' UTR start',\n",
       "       '3' UTR end', 'CDS Length', 'Strand', 'Transcription start site (TSS)',\n",
       "       'Transcript length (including UTRs and CDS)', 'Transcript start (bp)',\n",
       "       'Transcript end (bp)', 'Gene start (bp)', 'Gene end (bp)',\n",
       "       'Chromosome/scaffold name', 'CDS start', 'CDS end', 'Constitutive exon',\n",
       "       'Exon rank in transcript', 'Start phase', 'End phase',\n",
       "       'cDNA coding start', 'cDNA coding end', 'Genomic coding start',\n",
       "       'Genomic coding end', 'Exon stable ID', 'transcript_count', 'Miso_siso',\n",
       "       'Exon_length'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_miso_siso_to_keep.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "processed-comedy",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "nuclear-pleasure",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Exon region start (bp)</th>\n",
       "      <th>Exon region end (bp)</th>\n",
       "      <th>5' UTR start</th>\n",
       "      <th>5' UTR end</th>\n",
       "      <th>3' UTR start</th>\n",
       "      <th>3' UTR end</th>\n",
       "      <th>CDS start</th>\n",
       "      <th>CDS end</th>\n",
       "      <th>Miso_siso</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>20575776</td>\n",
       "      <td>20575887</td>\n",
       "      <td>20575776.0</td>\n",
       "      <td>20575871.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>20579608</td>\n",
       "      <td>20579691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>20582590</td>\n",
       "      <td>20582693</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>101.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>20584474</td>\n",
       "      <td>20584524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>20588024</td>\n",
       "      <td>20588105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536344</th>\n",
       "      <td>ENSG00000077585</td>\n",
       "      <td>ENST00000366592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>236178414</td>\n",
       "      <td>236178636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>465.0</td>\n",
       "      <td>687.0</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536345</th>\n",
       "      <td>ENSG00000077585</td>\n",
       "      <td>ENST00000366592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>236179879</td>\n",
       "      <td>236180028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>688.0</td>\n",
       "      <td>837.0</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536346</th>\n",
       "      <td>ENSG00000077585</td>\n",
       "      <td>ENST00000366592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>236183778</td>\n",
       "      <td>236183906</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>838.0</td>\n",
       "      <td>966.0</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536347</th>\n",
       "      <td>ENSG00000077585</td>\n",
       "      <td>ENST00000366592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>236205126</td>\n",
       "      <td>236205250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>967.0</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536348</th>\n",
       "      <td>ENSG00000077585</td>\n",
       "      <td>ENST00000366592</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>236208050</td>\n",
       "      <td>236208907</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>236208159.0</td>\n",
       "      <td>236208907.0</td>\n",
       "      <td>1092.0</td>\n",
       "      <td>1200.0</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>536349 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GeneID    transcript_id  Strand Chromosome/scaffold name  \\\n",
       "0       ENSG00000198692  ENST00000361365       1                        Y   \n",
       "1       ENSG00000198692  ENST00000361365       1                        Y   \n",
       "2       ENSG00000198692  ENST00000361365       1                        Y   \n",
       "3       ENSG00000198692  ENST00000361365       1                        Y   \n",
       "4       ENSG00000198692  ENST00000361365       1                        Y   \n",
       "...                 ...              ...     ...                      ...   \n",
       "536344  ENSG00000077585  ENST00000366592       1                        1   \n",
       "536345  ENSG00000077585  ENST00000366592       1                        1   \n",
       "536346  ENSG00000077585  ENST00000366592       1                        1   \n",
       "536347  ENSG00000077585  ENST00000366592       1                        1   \n",
       "536348  ENSG00000077585  ENST00000366592       1                        1   \n",
       "\n",
       "        Exon region start (bp)  Exon region end (bp)  5' UTR start  \\\n",
       "0                     20575776              20575887    20575776.0   \n",
       "1                     20579608              20579691           NaN   \n",
       "2                     20582590              20582693           NaN   \n",
       "3                     20584474              20584524           NaN   \n",
       "4                     20588024              20588105           NaN   \n",
       "...                        ...                   ...           ...   \n",
       "536344               236178414             236178636           NaN   \n",
       "536345               236179879             236180028           NaN   \n",
       "536346               236183778             236183906           NaN   \n",
       "536347               236205126             236205250           NaN   \n",
       "536348               236208050             236208907           NaN   \n",
       "\n",
       "        5' UTR end  3' UTR start   3' UTR end  CDS start  CDS end Miso_siso  \n",
       "0       20575871.0           NaN          NaN        1.0     16.0      Miso  \n",
       "1              NaN           NaN          NaN       17.0    100.0      Miso  \n",
       "2              NaN           NaN          NaN      101.0    204.0      Miso  \n",
       "3              NaN           NaN          NaN      205.0    255.0      Miso  \n",
       "4              NaN           NaN          NaN      256.0    337.0      Miso  \n",
       "...            ...           ...          ...        ...      ...       ...  \n",
       "536344         NaN           NaN          NaN      465.0    687.0      Siso  \n",
       "536345         NaN           NaN          NaN      688.0    837.0      Siso  \n",
       "536346         NaN           NaN          NaN      838.0    966.0      Siso  \n",
       "536347         NaN           NaN          NaN      967.0   1091.0      Siso  \n",
       "536348         NaN   236208159.0  236208907.0     1092.0   1200.0      Siso  \n",
       "\n",
       "[536349 rows x 13 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_tmp_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "grateful-reservation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "transcript_id\n",
       "ENST00000304952               [1000097.0]\n",
       "ENST00000327044                [959256.0]\n",
       "ENST00000338591                [960693.0]\n",
       "ENST00000341290      [981173.0, 982093.0]\n",
       "ENST00000379370               [1020172.0]\n",
       "ENST00000379407                [966531.0]\n",
       "ENST00000379409                [966531.0]\n",
       "ENST00000379410                [966531.0]\n",
       "ENST00000428771               [1000172.0]\n",
       "ENST00000484667                [999981.0]\n",
       "ENST00000620552    [1020373.0, 1022413.0]\n",
       "dtype: object"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_tmp_dev.sort_values(by=['Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'GeneID', 'transcript_id']).head(100).dropna(subset=[\"5' UTR end\"]).groupby('transcript_id').apply(lambda r: r[\"5' UTR end\"].values.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "confident-creativity",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "engine = create_engine(\"sqlite:////gstock/GeneIso/V2/GeneIso.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "optical-pavilion",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 47544/47544 [13:18<00:00, 59.53it/s] \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Exon region start (bp)</th>\n",
       "      <th>Exon region end (bp)</th>\n",
       "      <th>Ordinal_nb</th>\n",
       "      <th>Ordinal_nb_inverted</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100636608</td>\n",
       "      <td>100636806</td>\n",
       "      <td>1</td>\n",
       "      <td>-8</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635558</td>\n",
       "      <td>100635746</td>\n",
       "      <td>2</td>\n",
       "      <td>-7</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635178</td>\n",
       "      <td>100635252</td>\n",
       "      <td>3</td>\n",
       "      <td>-6</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100633931</td>\n",
       "      <td>100634029</td>\n",
       "      <td>4</td>\n",
       "      <td>-5</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100633405</td>\n",
       "      <td>100633539</td>\n",
       "      <td>5</td>\n",
       "      <td>-4</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536344</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88706752</td>\n",
       "      <td>88708450</td>\n",
       "      <td>26</td>\n",
       "      <td>-1</td>\n",
       "      <td>1698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536345</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23763004</td>\n",
       "      <td>23763134</td>\n",
       "      <td>1</td>\n",
       "      <td>-4</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536346</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23807826</td>\n",
       "      <td>23807939</td>\n",
       "      <td>2</td>\n",
       "      <td>-3</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536347</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23808719</td>\n",
       "      <td>23808814</td>\n",
       "      <td>3</td>\n",
       "      <td>-2</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536348</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23827070</td>\n",
       "      <td>23828868</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>1798</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>536349 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Miso_siso           GeneID    transcript_id  Strand  \\\n",
       "0           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "1           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "2           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "3           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "4           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "...          ...              ...              ...     ...   \n",
       "536344      Siso  ENSG00000287542  ENST00000512194       1   \n",
       "536345      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "536346      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "536347      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "536348      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "\n",
       "       Chromosome/scaffold name  Exon region start (bp)  Exon region end (bp)  \\\n",
       "0                             X               100636608             100636806   \n",
       "1                             X               100635558             100635746   \n",
       "2                             X               100635178             100635252   \n",
       "3                             X               100633931             100634029   \n",
       "4                             X               100633405             100633539   \n",
       "...                         ...                     ...                   ...   \n",
       "536344                        4                88706752              88708450   \n",
       "536345                       19                23763004              23763134   \n",
       "536346                       19                23807826              23807939   \n",
       "536347                       19                23808719              23808814   \n",
       "536348                       19                23827070              23828868   \n",
       "\n",
       "        Ordinal_nb  Ordinal_nb_inverted  Length  \n",
       "0                1                   -8     198  \n",
       "1                2                   -7     188  \n",
       "2                3                   -6      74  \n",
       "3                4                   -5      98  \n",
       "4                5                   -4     134  \n",
       "...            ...                  ...     ...  \n",
       "536344          26                   -1    1698  \n",
       "536345           1                   -4     130  \n",
       "536346           2                   -3     113  \n",
       "536347           3                   -2      95  \n",
       "536348           4                   -1    1798  \n",
       "\n",
       "[536349 rows x 10 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def exon_ord(df):\n",
    "    if df.Strand.values[0] == 1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'])\n",
    "\n",
    "    elif df.Strand.values[0] == -1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'], ascending=False)\n",
    "    \n",
    "    df['Ordinal_nb'] = list(range(1,df.shape[0]+1))\n",
    "    df['Length'] = df['Exon region end (bp)'] - df['Exon region start (bp)']\n",
    "    df['Ordinal_nb_inverted'] = np.array(list(range(1,df.shape[0] + 1))[::-1]) * -1\n",
    "\n",
    "    return df[['Miso_siso', 'GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'Ordinal_nb', 'Ordinal_nb_inverted', 'Length']]\n",
    "    \n",
    "\n",
    "exon_tmp_dev = exon_miso_siso_to_keep[['GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', \"5' UTR start\", \"5' UTR end\", \"3' UTR start\", \"3' UTR end\", 'CDS start', 'CDS end', 'Miso_siso']]\n",
    "exon_tmp_dev = exon_tmp_dev.sort_values(by=['Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'GeneID', 'transcript_id'])\n",
    "\n",
    "# exon_tmp_dev = exon_tmp_dev.loc[exon_tmp_dev['GeneID'].isin(exon_tmp_dev['GeneID'].unique().tolist()[:100])]\n",
    "\n",
    "exon_tmp_dev = exon_tmp_dev.groupby(['Miso_siso', 'GeneID', 'transcript_id']).progress_apply(exon_ord).reset_index(drop=True)\n",
    "\n",
    "\n",
    "\n",
    "exon_tmp_dev['Chromosome/scaffold name'] = exon_tmp_dev['Chromosome/scaffold name'].astype(str)\n",
    "\n",
    "# exon_tmp_dev.to_sql('Exons', engine, if_exists='replace')\n",
    "\n",
    "exon_tmp_dev.to_parquet('/gstock/GeneIso/V2/Exons.parquet')\n",
    "\n",
    "exon_tmp_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "moved-opening",
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_tmp_dev = pd.read_parquet('/gstock/GeneIso/V2/Exons.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "assigned-westminster",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 47544/47544 [26:10<00:00, 30.27it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Exon region start (bp)</th>\n",
       "      <th>Exon region end (bp)</th>\n",
       "      <th>5' UTR start</th>\n",
       "      <th>5' UTR end</th>\n",
       "      <th>3' UTR start</th>\n",
       "      <th>3' UTR end</th>\n",
       "      <th>CDS start</th>\n",
       "      <th>CDS end</th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>Ordinal_nb</th>\n",
       "      <th>Ordinal_nb_inverted</th>\n",
       "      <th>CDS_start_abs</th>\n",
       "      <th>CDS_end_abs</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100636608</td>\n",
       "      <td>100636806</td>\n",
       "      <td>100636695.0</td>\n",
       "      <td>100636806.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>1</td>\n",
       "      <td>-7</td>\n",
       "      <td>100636608.0</td>\n",
       "      <td>100636694.0</td>\n",
       "      <td>86.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635558</td>\n",
       "      <td>100635746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>2</td>\n",
       "      <td>-6</td>\n",
       "      <td>100635558.0</td>\n",
       "      <td>100635746.0</td>\n",
       "      <td>188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635178</td>\n",
       "      <td>100635252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>277.0</td>\n",
       "      <td>351.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>3</td>\n",
       "      <td>-5</td>\n",
       "      <td>100635178.0</td>\n",
       "      <td>100635252.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100633931</td>\n",
       "      <td>100634029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>352.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>4</td>\n",
       "      <td>-4</td>\n",
       "      <td>100633931.0</td>\n",
       "      <td>100634029.0</td>\n",
       "      <td>98.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100633405</td>\n",
       "      <td>100633539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>451.0</td>\n",
       "      <td>585.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>5</td>\n",
       "      <td>-3</td>\n",
       "      <td>100633405.0</td>\n",
       "      <td>100633539.0</td>\n",
       "      <td>134.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497254</th>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88704098</td>\n",
       "      <td>88704281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2634.0</td>\n",
       "      <td>2817.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>21</td>\n",
       "      <td>-3</td>\n",
       "      <td>88704098.0</td>\n",
       "      <td>88704281.0</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497255</th>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88704508</td>\n",
       "      <td>88704610</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2818.0</td>\n",
       "      <td>2920.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>22</td>\n",
       "      <td>-2</td>\n",
       "      <td>88704508.0</td>\n",
       "      <td>88704610.0</td>\n",
       "      <td>102.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497256</th>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88706752</td>\n",
       "      <td>88708450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88706961.0</td>\n",
       "      <td>88708450.0</td>\n",
       "      <td>2921.0</td>\n",
       "      <td>3129.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>23</td>\n",
       "      <td>-1</td>\n",
       "      <td>88706752.0</td>\n",
       "      <td>88706960.0</td>\n",
       "      <td>208.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497257</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23827070</td>\n",
       "      <td>23828868</td>\n",
       "      <td>23827070.0</td>\n",
       "      <td>23827161.0</td>\n",
       "      <td>23828050.0</td>\n",
       "      <td>23828868.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>888.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>23827162.0</td>\n",
       "      <td>23828868.0</td>\n",
       "      <td>887.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497258</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23827070</td>\n",
       "      <td>23828868</td>\n",
       "      <td>23827070.0</td>\n",
       "      <td>23827161.0</td>\n",
       "      <td>23828050.0</td>\n",
       "      <td>23828868.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>888.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>23827070.0</td>\n",
       "      <td>23828049.0</td>\n",
       "      <td>887.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>497259 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GeneID    transcript_id  Strand Chromosome/scaffold name  \\\n",
       "0       ENSG00000000003  ENST00000373020      -1                        X   \n",
       "1       ENSG00000000003  ENST00000373020      -1                        X   \n",
       "2       ENSG00000000003  ENST00000373020      -1                        X   \n",
       "3       ENSG00000000003  ENST00000373020      -1                        X   \n",
       "4       ENSG00000000003  ENST00000373020      -1                        X   \n",
       "...                 ...              ...     ...                      ...   \n",
       "497254  ENSG00000287542  ENST00000512194       1                        4   \n",
       "497255  ENSG00000287542  ENST00000512194       1                        4   \n",
       "497256  ENSG00000287542  ENST00000512194       1                        4   \n",
       "497257  ENSG00000288920  ENST00000484897       1                       19   \n",
       "497258  ENSG00000288920  ENST00000484897       1                       19   \n",
       "\n",
       "        Exon region start (bp)  Exon region end (bp)  5' UTR start  \\\n",
       "0                    100636608             100636806   100636695.0   \n",
       "1                    100635558             100635746           NaN   \n",
       "2                    100635178             100635252           NaN   \n",
       "3                    100633931             100634029           NaN   \n",
       "4                    100633405             100633539           NaN   \n",
       "...                        ...                   ...           ...   \n",
       "497254                88704098              88704281           NaN   \n",
       "497255                88704508              88704610           NaN   \n",
       "497256                88706752              88708450           NaN   \n",
       "497257                23827070              23828868    23827070.0   \n",
       "497258                23827070              23828868    23827070.0   \n",
       "\n",
       "         5' UTR end  3' UTR start  3' UTR end  CDS start  CDS end Miso_siso  \\\n",
       "0       100636806.0           NaN         NaN        1.0     87.0      Miso   \n",
       "1               NaN           NaN         NaN       88.0    276.0      Miso   \n",
       "2               NaN           NaN         NaN      277.0    351.0      Miso   \n",
       "3               NaN           NaN         NaN      352.0    450.0      Miso   \n",
       "4               NaN           NaN         NaN      451.0    585.0      Miso   \n",
       "...             ...           ...         ...        ...      ...       ...   \n",
       "497254          NaN           NaN         NaN     2634.0   2817.0      Siso   \n",
       "497255          NaN           NaN         NaN     2818.0   2920.0      Siso   \n",
       "497256          NaN    88706961.0  88708450.0     2921.0   3129.0      Siso   \n",
       "497257   23827161.0    23828050.0  23828868.0        1.0    888.0      Siso   \n",
       "497258   23827161.0    23828050.0  23828868.0        1.0    888.0      Siso   \n",
       "\n",
       "        Ordinal_nb  Ordinal_nb_inverted  CDS_start_abs  CDS_end_abs  Length  \n",
       "0                1                   -7    100636608.0  100636694.0    86.0  \n",
       "1                2                   -6    100635558.0  100635746.0   188.0  \n",
       "2                3                   -5    100635178.0  100635252.0    74.0  \n",
       "3                4                   -4    100633931.0  100634029.0    98.0  \n",
       "4                5                   -3    100633405.0  100633539.0   134.0  \n",
       "...            ...                  ...            ...          ...     ...  \n",
       "497254          21                   -3     88704098.0   88704281.0   183.0  \n",
       "497255          22                   -2     88704508.0   88704610.0   102.0  \n",
       "497256          23                   -1     88706752.0   88706960.0   208.0  \n",
       "497257           1                   -1     23827162.0   23828868.0   887.0  \n",
       "497258           1                   -1     23827070.0   23828049.0   887.0  \n",
       "\n",
       "[497259 rows x 18 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def retrieve_cds_coord(df):\n",
    "    if df.Strand.values[0] == 1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'])\n",
    "\n",
    "        df_cds = df.dropna(subset=['CDS start', 'CDS end'])\n",
    "        df_cds['Ordinal_nb'] = list(range(1,df_cds.shape[0]+1))\n",
    "        df_cds['Ordinal_nb_inverted'] = np.array(list(range(1,df_cds.shape[0] + 1))[::-1]) * -1\n",
    "\n",
    "        df_cds_start = df_cds.loc[df_cds['Ordinal_nb'] == 1]\n",
    "        df_cds_start['CDS_start_abs'] = df_cds_start[\"5' UTR end\"] + 1\n",
    "        df_cds_start['CDS_end_abs'] = df_cds_start[\"Exon region end (bp)\"]\n",
    "\n",
    "        df_cds_core = df_cds.loc[~df_cds['Ordinal_nb'].isin([1, df_cds.shape[0]])]\n",
    "        df_cds_core['CDS_start_abs'] = df_cds_core['Exon region start (bp)'] \n",
    "        df_cds_core['CDS_end_abs'] = df_cds_core['Exon region end (bp)'] \n",
    "\n",
    "        df_cds_end = df_cds.loc[df_cds['Ordinal_nb'] == df_cds.shape[0]]\n",
    "        df_cds_end['CDS_start_abs'] = df_cds_end['Exon region start (bp)'] \n",
    "        df_cds_end['CDS_end_abs'] = df_cds_end[\"3' UTR start\"] - 1\n",
    "\n",
    "        df_cds_final = pd.concat([df_cds_start, df_cds_core, df_cds_end])\n",
    "\n",
    "    elif df.Strand.values[0] == -1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'], ascending=False)\n",
    "\n",
    "        df_cds = df.dropna(subset=['CDS start', 'CDS end'])\n",
    "        df_cds['Ordinal_nb'] = list(range(1,df_cds.shape[0]+1))\n",
    "        df_cds['Ordinal_nb_inverted'] = np.array(list(range(1,df_cds.shape[0] + 1))[::-1]) * -1\n",
    "\n",
    "        \n",
    "        df_cds_start = df_cds.loc[df_cds['Ordinal_nb'] == 1]\n",
    "        df_cds_start['CDS_start_abs'] = df_cds_start[\"Exon region start (bp)\"]\n",
    "        df_cds_start['CDS_end_abs'] = df_cds_start[\"5' UTR start\"] - 1\n",
    "\n",
    "        df_cds_core = df_cds.loc[~df_cds['Ordinal_nb'].isin([1, df_cds.shape[0]])]\n",
    "        df_cds_core['CDS_start_abs'] = df_cds_core['Exon region start (bp)'] \n",
    "        df_cds_core['CDS_end_abs'] = df_cds_core['Exon region end (bp)'] \n",
    "\n",
    "        df_cds_end = df_cds.loc[df_cds['Ordinal_nb'] == df_cds.shape[0]]\n",
    "        df_cds_end['CDS_start_abs'] = df_cds_end[\"3' UTR end\"] + 1\n",
    "        df_cds_end['CDS_end_abs'] = df_cds_end['Exon region end (bp)']\n",
    "\n",
    "        df_cds_final = pd.concat([df_cds_start, df_cds_core, df_cds_end])\n",
    "\n",
    "    df_cds_final['Length'] = df_cds_final['CDS end'] - df_cds_final['CDS start']\n",
    "    return df_cds_final\n",
    "\n",
    "\n",
    "exon_tmp_dev = exon_miso_siso_to_keep[['GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', \"5' UTR start\", \"5' UTR end\", \"3' UTR start\", \"3' UTR end\", 'CDS start', 'CDS end', 'Miso_siso']]\n",
    "exon_tmp_dev = exon_tmp_dev.sort_values(by=['Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'GeneID', 'transcript_id'])\n",
    "exon_tmp_dev['Chromosome/scaffold name'] = exon_tmp_dev['Chromosome/scaffold name'].astype(str)\n",
    "\n",
    "# exon_tmp_dev = exon_tmp_dev.loc[exon_tmp_dev['GeneID'].isin(exon_tmp_dev['GeneID'].unique().tolist()[:100])]\n",
    "\n",
    "cds = exon_tmp_dev.groupby(['Miso_siso', 'GeneID', 'transcript_id']).progress_apply(retrieve_cds_coord).reset_index(drop=True)\n",
    "\n",
    "# cds.to_sql('CDS', engine, if_exists='replace')\n",
    "cds.to_parquet('/gstock/GeneIso/V2/CDS.parquet')\n",
    "\n",
    "cds\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "following-acrobat",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 47544/47544 [14:46<00:00, 53.60it/s] \n",
      "100%|██████████| 47544/47544 [14:44<00:00, 53.76it/s] \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Exon region start (bp)</th>\n",
       "      <th>Exon region end (bp)</th>\n",
       "      <th>5' UTR start</th>\n",
       "      <th>5' UTR end</th>\n",
       "      <th>3' UTR start</th>\n",
       "      <th>3' UTR end</th>\n",
       "      <th>CDS start</th>\n",
       "      <th>CDS end</th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>Ordinal_nb</th>\n",
       "      <th>Ordinal_nb_inverted</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100636608</td>\n",
       "      <td>100636806</td>\n",
       "      <td>100636695.0</td>\n",
       "      <td>100636806.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>111.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000612152</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100636793</td>\n",
       "      <td>100637104</td>\n",
       "      <td>100636793.0</td>\n",
       "      <td>100637104.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Miso</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>311.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000612152</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635558</td>\n",
       "      <td>100635746</td>\n",
       "      <td>100635570.0</td>\n",
       "      <td>100635746.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>176.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000614008</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100636793</td>\n",
       "      <td>100637104</td>\n",
       "      <td>100636793.0</td>\n",
       "      <td>100637104.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Miso</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>311.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000614008</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635558</td>\n",
       "      <td>100635746</td>\n",
       "      <td>100635570.0</td>\n",
       "      <td>100635746.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Miso</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>176.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82355</th>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88605795</td>\n",
       "      <td>88606049</td>\n",
       "      <td>88605795.0</td>\n",
       "      <td>88605823.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>226.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82356</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23763004</td>\n",
       "      <td>23763134</td>\n",
       "      <td>23763004.0</td>\n",
       "      <td>23763134.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Siso</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>130.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82357</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23807826</td>\n",
       "      <td>23807939</td>\n",
       "      <td>23807826.0</td>\n",
       "      <td>23807939.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Siso</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>113.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82358</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23808719</td>\n",
       "      <td>23808814</td>\n",
       "      <td>23808719.0</td>\n",
       "      <td>23808814.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Siso</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82359</th>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23827070</td>\n",
       "      <td>23828868</td>\n",
       "      <td>23827070.0</td>\n",
       "      <td>23827161.0</td>\n",
       "      <td>23828050.0</td>\n",
       "      <td>23828868.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>888.0</td>\n",
       "      <td>Siso</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82360 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID    transcript_id  Strand Chromosome/scaffold name  \\\n",
       "0      ENSG00000000003  ENST00000373020      -1                        X   \n",
       "1      ENSG00000000003  ENST00000612152      -1                        X   \n",
       "2      ENSG00000000003  ENST00000612152      -1                        X   \n",
       "3      ENSG00000000003  ENST00000614008      -1                        X   \n",
       "4      ENSG00000000003  ENST00000614008      -1                        X   \n",
       "...                ...              ...     ...                      ...   \n",
       "82355  ENSG00000287542  ENST00000512194       1                        4   \n",
       "82356  ENSG00000288920  ENST00000484897       1                       19   \n",
       "82357  ENSG00000288920  ENST00000484897       1                       19   \n",
       "82358  ENSG00000288920  ENST00000484897       1                       19   \n",
       "82359  ENSG00000288920  ENST00000484897       1                       19   \n",
       "\n",
       "       Exon region start (bp)  Exon region end (bp)  5' UTR start  \\\n",
       "0                   100636608             100636806   100636695.0   \n",
       "1                   100636793             100637104   100636793.0   \n",
       "2                   100635558             100635746   100635570.0   \n",
       "3                   100636793             100637104   100636793.0   \n",
       "4                   100635558             100635746   100635570.0   \n",
       "...                       ...                   ...           ...   \n",
       "82355                88605795              88606049    88605795.0   \n",
       "82356                23763004              23763134    23763004.0   \n",
       "82357                23807826              23807939    23807826.0   \n",
       "82358                23808719              23808814    23808719.0   \n",
       "82359                23827070              23828868    23827070.0   \n",
       "\n",
       "        5' UTR end  3' UTR start  3' UTR end  CDS start  CDS end Miso_siso  \\\n",
       "0      100636806.0           NaN         NaN        1.0     87.0      Miso   \n",
       "1      100637104.0           NaN         NaN        NaN      NaN      Miso   \n",
       "2      100635746.0           NaN         NaN        1.0     12.0      Miso   \n",
       "3      100637104.0           NaN         NaN        NaN      NaN      Miso   \n",
       "4      100635746.0           NaN         NaN        1.0     12.0      Miso   \n",
       "...            ...           ...         ...        ...      ...       ...   \n",
       "82355   88605823.0           NaN         NaN        1.0    226.0      Siso   \n",
       "82356   23763134.0           NaN         NaN        NaN      NaN      Siso   \n",
       "82357   23807939.0           NaN         NaN        NaN      NaN      Siso   \n",
       "82358   23808814.0           NaN         NaN        NaN      NaN      Siso   \n",
       "82359   23827161.0    23828050.0  23828868.0        1.0    888.0      Siso   \n",
       "\n",
       "       Ordinal_nb  Ordinal_nb_inverted  Length  \n",
       "0             1.0                 -1.0   111.0  \n",
       "1             1.0                 -2.0   311.0  \n",
       "2             2.0                 -1.0   176.0  \n",
       "3             1.0                 -2.0   311.0  \n",
       "4             2.0                 -1.0   176.0  \n",
       "...           ...                  ...     ...  \n",
       "82355         4.0                 -1.0    28.0  \n",
       "82356         1.0                 -4.0   130.0  \n",
       "82357         2.0                 -3.0   113.0  \n",
       "82358         3.0                 -2.0    95.0  \n",
       "82359         4.0                 -1.0    91.0  \n",
       "\n",
       "[82360 rows x 16 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def retrieve_5_utr(df):\n",
    "    if df.Strand.values[0] == 1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'])\n",
    "\n",
    "    elif df.Strand.values[0] == -1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'], ascending=False)\n",
    "\n",
    "    df_5_utr = df.dropna(subset=[\"5' UTR start\", \"5' UTR end\"])\n",
    "    df_5_utr['Ordinal_nb'] = list(range(1,df_5_utr.shape[0]+1))\n",
    "    df_5_utr['Ordinal_nb_inverted'] = np.array(list(range(1,df_5_utr.shape[0] + 1))[::-1]) * -1\n",
    "    \n",
    "    df_5_utr['Length'] = df_5_utr[\"5' UTR end\"] - df_5_utr[\"5' UTR start\"]\n",
    "    return df_5_utr\n",
    "\n",
    "def retrieve_3_utr(df):\n",
    "    if df.Strand.values[0] == 1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'])\n",
    "\n",
    "    elif df.Strand.values[0] == -1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'], ascending=False)\n",
    "\n",
    "    df_3_utr = df.dropna(subset=[\"3' UTR start\", \"3' UTR end\"])\n",
    "    df_3_utr['Ordinal_nb'] = list(range(1,df_3_utr.shape[0]+1))\n",
    "    df_3_utr['Ordinal_nb_inverted'] = np.array(list(range(1,df_3_utr.shape[0] + 1))[::-1]) * -1\n",
    "    \n",
    "    df_3_utr['Length'] = df_3_utr[\"3' UTR end\"] - df_3_utr[\"3' UTR start\"]\n",
    "\n",
    "    return df_3_utr\n",
    "\n",
    "exon_tmp_dev = exon_miso_siso_to_keep[['GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', \"5' UTR start\", \"5' UTR end\", \"3' UTR start\", \"3' UTR end\", 'CDS start', 'CDS end', 'Miso_siso']]\n",
    "exon_tmp_dev = exon_tmp_dev.sort_values(by=['Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'GeneID', 'transcript_id'])\n",
    "\n",
    "# exon_tmp_dev = exon_tmp_dev.loc[exon_tmp_dev['GeneID'].isin(exon_tmp_dev['GeneID'].unique().tolist()[:100])]\n",
    "\n",
    "df_5_utr = exon_tmp_dev.groupby(['Miso_siso', 'GeneID', 'transcript_id']).progress_apply(retrieve_5_utr).reset_index(drop=True)\n",
    "df_3_utr = exon_tmp_dev.groupby(['Miso_siso', 'GeneID', 'transcript_id']).progress_apply(retrieve_3_utr).reset_index(drop=True)\n",
    "\n",
    "\n",
    "df_5_utr['Chromosome/scaffold name'] = df_5_utr['Chromosome/scaffold name'].astype(str)\n",
    "df_3_utr['Chromosome/scaffold name'] = df_3_utr['Chromosome/scaffold name'].astype(str)\n",
    "\n",
    "\n",
    "# df_5_utr.to_sql('5_UTR', engine, if_exists='replace')\n",
    "df_5_utr.to_parquet('/gstock/GeneIso/V2/5_UTR.parquet')\n",
    "\n",
    "# df_3_utr.to_sql('3_UTR', engine, if_exists='replace')\n",
    "df_3_utr.to_parquet('/gstock/GeneIso/V2/3_UTR.parquet')\n",
    "\n",
    "df_5_utr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "finite-phenomenon",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 47544/47544 [17:01<00:00, 46.53it/s] \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Introns</th>\n",
       "      <th>Ordinal_nb</th>\n",
       "      <th>Ordinal_nb_inverted</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635747-100636607</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-7</td>\n",
       "      <td>860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100635253-100635557</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-6</td>\n",
       "      <td>304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100634030-100635177</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-5</td>\n",
       "      <td>1147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100633540-100633930</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-4</td>\n",
       "      <td>390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Miso</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>-1</td>\n",
       "      <td>X</td>\n",
       "      <td>100632569-100633404</td>\n",
       "      <td>5.0</td>\n",
       "      <td>-3</td>\n",
       "      <td>835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488800</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88704282-88704507</td>\n",
       "      <td>24.0</td>\n",
       "      <td>-3</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488801</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000287542</td>\n",
       "      <td>ENST00000512194</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>88704611-88706751</td>\n",
       "      <td>25.0</td>\n",
       "      <td>-2</td>\n",
       "      <td>2140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488802</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23763135-23807825</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-4</td>\n",
       "      <td>44690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488803</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23807940-23808718</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-3</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488804</th>\n",
       "      <td>Siso</td>\n",
       "      <td>ENSG00000288920</td>\n",
       "      <td>ENST00000484897</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>23808815-23827069</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-2</td>\n",
       "      <td>18254</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>488805 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Miso_siso           GeneID    transcript_id  Strand  \\\n",
       "0           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "1           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "2           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "3           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "4           Miso  ENSG00000000003  ENST00000373020      -1   \n",
       "...          ...              ...              ...     ...   \n",
       "488800      Siso  ENSG00000287542  ENST00000512194       1   \n",
       "488801      Siso  ENSG00000287542  ENST00000512194       1   \n",
       "488802      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "488803      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "488804      Siso  ENSG00000288920  ENST00000484897       1   \n",
       "\n",
       "       Chromosome/scaffold name              Introns  Ordinal_nb  \\\n",
       "0                             X  100635747-100636607         1.0   \n",
       "1                             X  100635253-100635557         2.0   \n",
       "2                             X  100634030-100635177         3.0   \n",
       "3                             X  100633540-100633930         4.0   \n",
       "4                             X  100632569-100633404         5.0   \n",
       "...                         ...                  ...         ...   \n",
       "488800                        4    88704282-88704507        24.0   \n",
       "488801                        4    88704611-88706751        25.0   \n",
       "488802                       19    23763135-23807825         1.0   \n",
       "488803                       19    23807940-23808718         2.0   \n",
       "488804                       19    23808815-23827069         3.0   \n",
       "\n",
       "        Ordinal_nb_inverted  Length  \n",
       "0                        -7     860  \n",
       "1                        -6     304  \n",
       "2                        -5    1147  \n",
       "3                        -4     390  \n",
       "4                        -3     835  \n",
       "...                     ...     ...  \n",
       "488800                   -3     225  \n",
       "488801                   -2    2140  \n",
       "488802                   -4   44690  \n",
       "488803                   -3     778  \n",
       "488804                   -2   18254  \n",
       "\n",
       "[488805 rows x 9 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def exon_ord(df):\n",
    "\n",
    "    \n",
    "    df['Exon_nb'] = list(range(1,df.shape[0]+1))\n",
    "\n",
    "    return df\n",
    "    \n",
    "\n",
    "# Fct to compute intron boundaries\n",
    "def get_introns(df):\n",
    "    \n",
    "#     if df.Strand.values[0] == 1:\n",
    "    df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'])\n",
    "\n",
    "    l = list()\n",
    "    exons = df['ranges'].values.tolist()\n",
    "    for j, e in enumerate(exons):\n",
    "        # Exon 1\n",
    "        if j == 0:\n",
    "            l.append(int(e.split(\"-\")[1]) + 1)\n",
    "            \n",
    "        # Exon 2 <-> Exon -2\n",
    "        elif j > 0 and j < len(exons) - 1:\n",
    "            l.append(int(e.split(\"-\")[0]) - 1)\n",
    "            l.append(int(e.split(\"-\")[1]) + 1)\n",
    "            \n",
    "        # Exon -1\n",
    "        elif j == len(exons) - 1:\n",
    "            l.append(int(e.split(\"-\")[0]) - 1)\n",
    "\n",
    "    # Final list        \n",
    "    l = [\"{}-{}\".format(e, l[j + 1]) for j, e in enumerate(l) if j < len(l) - 1 if j % 2 == 0]\n",
    "    df['Introns'] = l + [np.nan]\n",
    "\n",
    "    \n",
    "    if df.Strand.values[0] == -1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'], ascending=False)\n",
    "\n",
    "        df.loc[df['Introns'].isna() == False,  'Ordinal_nb'] = range(1,len(l)+1) \n",
    "        df['Ordinal_nb_inverted'] = np.array(list(range(1,df.shape[0] + 1))[::-1]) * -1\n",
    "    \n",
    "    if df.Strand.values[0] == 1:\n",
    "        df = df.sort_values(by=['Exon region start (bp)', 'Exon region end (bp)'])\n",
    "\n",
    "        df.loc[df['Introns'].isna() == False,  'Ordinal_nb'] = range(1,len(l)+1) \n",
    "        df['Ordinal_nb_inverted'] = (np.array(list(range(1,df.shape[0] + 1))[::-1]) * -1) + 1\n",
    "\n",
    "#         df['Introns_nb'] = range(1,len(l)+2)    \n",
    "\n",
    "    return df[['Miso_siso', 'GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Introns', 'Ordinal_nb', 'Ordinal_nb_inverted']].dropna()\n",
    "#     return df[['GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Introns', 'Introns_nb']]\n",
    "\n",
    "\n",
    "exon_tmp_dev = exon_miso_siso_to_keep[['GeneID', 'transcript_id', 'Strand', 'Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', \"5' UTR start\", \"5' UTR end\", \"3' UTR start\", \"3' UTR end\", 'CDS start', 'CDS end', 'Miso_siso']]\n",
    "exon_tmp_dev = exon_tmp_dev.sort_values(by=['Chromosome/scaffold name', 'Exon region start (bp)', 'Exon region end (bp)', 'GeneID', 'transcript_id'])\n",
    "exon_tmp_dev['ranges'] = exon_tmp_dev['Exon region start (bp)'].astype(str) + '-' + exon_tmp_dev['Exon region end (bp)'].astype(str)\n",
    "\n",
    "# exon_tmp_dev = exon_tmp_dev.loc[exon_tmp_dev['GeneID'].isin(exon_tmp_dev['GeneID'].unique().tolist()[:100])]\n",
    "\n",
    "\n",
    "\n",
    "introns_df = exon_tmp_dev.groupby(['Miso_siso', 'GeneID', 'transcript_id']).progress_apply(get_introns).reset_index(drop=True)\n",
    "introns_df['Chromosome/scaffold name'] = introns_df['Chromosome/scaffold name'].astype(str)\n",
    "introns_df[\"Length\"] = introns_df[\"Introns\"].apply(lambda r: int(r.split(\"-\")[1]) - int(r.split(\"-\")[0]))\n",
    "\n",
    "\n",
    "# introns_df.to_sql('Introns', engine, if_exists='replace')\n",
    "introns_df.to_parquet('/gstock/GeneIso/V2/Introns.parquet')\n",
    "\n",
    "introns_df\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "6bc58787a7e3b85efa746fe1d312329650c66964b196f4af9721c9b0ded7fae1"
  },
  "kernelspec": {
   "display_name": "Python [conda env:.conda-ExoCarto] *",
   "language": "python",
   "name": "conda-env-.conda-ExoCarto-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
