{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "requested-waste",
   "metadata": {},
   "source": [
    "# Imports & Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "informed-split",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/weber/.conda/envs/ExoCarto/lib/python3.7/site-packages/ipykernel_launcher.py:38: MatplotlibDeprecationWarning: \n",
      "The createFontList function was deprecated in Matplotlib 3.2 and will be removed two minor releases later. Use FontManager.addfont instead.\n"
     ]
    }
   ],
   "source": [
    "# General imports\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import scipy\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "sys.path.append('../')\n",
    "# Other imports\n",
    "\n",
    "tqdm.pandas()\n",
    "import yaml\n",
    "import json\n",
    "\n",
    "# Figures imports\n",
    "import matplotlib\n",
    "from matplotlib.lines import Line2D   \n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as mtick\n",
    "from statannot import add_stat_annotation\n",
    "import matplotlib.font_manager as font_manager\n",
    "import string\n",
    "\n",
    "\n",
    "# Font settings\n",
    "font_dirs = ['/home/weber/Fonts', ]\n",
    "font_files = font_manager.findSystemFonts(fontpaths=font_dirs)\n",
    "font_list = font_manager.createFontList(font_files)\n",
    "font_manager.fontManager.ttflist.extend(font_list)\n",
    "\n",
    "pvalues_cutoff = json.load(open('config/config_files.json'))['pvalues_cutoff']\n",
    "pvalues_cutoff = [[e, str(e)] for e in pvalues_cutoff]\n",
    "\n",
    "import matplotlib.font_manager as font_manager\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['font.family'] = 'sans-serif'\n",
    "rcParams['font.sans-serif'] = ['Arial']\n",
    "rcParams['font.weight'] = 'light'\n",
    "rcParams['axes.facecolor'] = 'white'\n",
    "rcParams['font.size'] = 18"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "photographic-matthew",
   "metadata": {},
   "source": [
    "# Load files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "placed-throat",
   "metadata": {},
   "source": [
    "## Homo sapiens"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "international-orange",
   "metadata": {},
   "source": [
    "### Update with count & write file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ethical-activity",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████| 13461/13461 [00:30<00:00, 440.79it/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████| 13478/13478 [00:30<00:00, 438.89it/s]\n"
     ]
    }
   ],
   "source": [
    "# H sapiens\n",
    "\n",
    "hs_genes = pd.read_parquet(\"/gstock/GeneIso/V2/Genes.parquet\")\n",
    "hs_mrna = pd.read_parquet(\"/gstock/GeneIso/V2/mRNA.parquet\")\n",
    "hs_exons = pd.read_parquet(\"/gstock/GeneIso/V2/Exons.parquet\")\n",
    "hs_exons['ranges'] = hs_exons['Exon region start (bp)'].astype(str) + '-' + hs_exons['Exon region end (bp)'].astype(str)\n",
    "\n",
    "hs_cds = pd.read_parquet(\"/gstock/GeneIso/V2/CDS.parquet\")\n",
    "hs_cds['ranges'] = hs_cds['Exon region start (bp)'].astype(str) + '-' + hs_cds['Exon region end (bp)'].astype(str)\n",
    "\n",
    "# TODO > REMOVE\n",
    "## Used to compute nb of completely different exon boundaries / gene => differ from exon events where overlap is possible\n",
    "hs_cds = hs_cds.groupby('GeneID').progress_apply(overlap_apply)\n",
    "hs_cds.to_parquet(\"/gstock/GeneIso/V2/CDS_update.parquet\")\n",
    "hs_exons = hs_exons.groupby('GeneID').progress_apply(overlap_apply)\n",
    "hs_exons.to_parquet(\"/gstock/GeneIso/V2/Exons_update.parquet\")\n",
    "\n",
    "hs_five_UTR = pd.read_parquet(\"/gstock/GeneIso/V2/5_UTR.parquet\")\n",
    "hs_three_UTR = pd.read_parquet(\"/gstock/GeneIso/V2/3_UTR.parquet\")\n",
    "\n",
    "# TODO > REMOVE\n",
    "hs_five_UTR['ranges']  = hs_five_UTR[\"5' UTR start\"].astype(int).astype(str) + '-' +  hs_five_UTR[\"5' UTR end\"].astype(int).astype(str)\n",
    "hs_three_UTR['ranges'] = hs_three_UTR[\"3' UTR start\"].astype(int).astype(str) + '-' + hs_three_UTR[\"3' UTR end\"].astype(int).astype(str)\n",
    "hs_five_UTR = hs_five_UTR.groupby('GeneID').progress_apply(overlap_apply)\n",
    "hs_five_UTR.to_parquet(\"/gstock/GeneIso/V2/5_UTR_update.parquet\")\n",
    "\n",
    "hs_three_UTR = hs_three_UTR.groupby('GeneID').progress_apply(overlap_apply)\n",
    "hs_three_UTR.to_parquet(\"/gstock/GeneIso/V2/3_UTR_update.parquet\")\n",
    "\n",
    "hs_introns = pd.read_parquet(\"/gstock/GeneIso/V2/Introns.parquet\")\n",
    "hs_introns.loc[hs_introns['Strand'] == 1, 'Ordinal_nb_inverted'] = hs_introns.loc[hs_introns['Strand'] == 1, 'Ordinal_nb_inverted'] + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "endangered-mexico",
   "metadata": {},
   "source": [
    "## Mus Musculus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "incomplete-certification",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████| 12154/12154 [00:27<00:00, 437.20it/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████| 12154/12154 [00:24<00:00, 504.86it/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████| 12099/12099 [00:29<00:00, 404.02it/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████| 12130/12130 [00:29<00:00, 404.95it/s]\n"
     ]
    }
   ],
   "source": [
    "genes = pd.read_parquet(\"/gstock/GeneIso/V2/Genes_MM.parquet\")\n",
    "# mrna = pd.read_parquet(\"/gstock/GeneIso/V2/mRNA.parquet\")\n",
    "exons = pd.read_parquet(\"/gstock/GeneIso/V2/Exons_MM.parquet\")\n",
    "exons['ranges'] = exons['Exon region start (bp)'].astype(str) + '-' + exons['Exon region end (bp)'].astype(str)\n",
    "\n",
    "\n",
    "cds = pd.read_parquet(\"/gstock/GeneIso/V2/CDS_MM.parquet\")\n",
    "cds['ranges'] = cds['Exon region start (bp)'].astype(str) + '-' + cds['Exon region end (bp)'].astype(str)\n",
    "\n",
    "# TODO > REMOVE\n",
    "cds = cds.groupby('GeneID').progress_apply(overlap_apply)\n",
    "exons = exons.groupby('GeneID').progress_apply(overlap_apply)\n",
    "\n",
    "cds.to_parquet(\"/gstock/GeneIso/V2/CDS_update_MM.parquet\")\n",
    "exons.to_parquet(\"/gstock/GeneIso/V2/Exons_update_MM.parquet\")\n",
    "\n",
    "\n",
    "five_UTR = pd.read_parquet(\"/gstock/GeneIso/V2/5_UTR_MM.parquet\")\n",
    "three_UTR = pd.read_parquet(\"/gstock/GeneIso/V2/3_UTR_MM.parquet\")\n",
    "introns = pd.read_parquet(\"/gstock/GeneIso/V2/Introns_MM.parquet\")\n",
    "\n",
    "# TODO > REMOVE\n",
    "## Used to compute nb of completely different exon boundaries / gene => differ from exon events where overlap is possible\n",
    "\n",
    "five_UTR['ranges'] = five_UTR['UTR_start'].astype(int).astype(str) + '-' + five_UTR['UTR_end'].astype(int).astype(str)\n",
    "three_UTR['ranges'] = three_UTR['UTR_start'].astype(int).astype(str) + '-' + three_UTR['UTR_end'].astype(int).astype(str)\n",
    "five_UTR = five_UTR.groupby('GeneID').progress_apply(overlap_apply)\n",
    "three_UTR = three_UTR.groupby('GeneID').progress_apply(overlap_apply)\n",
    "\n",
    "five_UTR.to_parquet(\"/gstock/GeneIso/V2/5_UTR_update_MM.parquet\")\n",
    "three_UTR.to_parquet(\"/gstock/GeneIso/V2/3_UTR_update_MM.parquet\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#TODO\n",
    "# introns.loc[introns['Strand'] == 1, 'Ordinal_nb_inverted'] = introns.loc[introns['Strand'] == 1, 'Ordinal_nb_inverted'] + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "demanding-crash",
   "metadata": {},
   "source": [
    "# Descritive stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faced-story",
   "metadata": {},
   "source": [
    "## Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "marine-shanghai",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_new = genes[['Miso_siso', 'GeneID', 'Gene name', 'ensembl_gene_id', 'chromosome_name', 'start_position', 'end_position']].drop_duplicates()\n",
    "genes_new['Gene_length'] = genes_new['end_position'] - genes_new['start_position']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "atlantic-stage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">HS</th>\n",
       "      <th>MISOG</th>\n",
       "      <td>9826</td>\n",
       "      <td>82044</td>\n",
       "      <td>143420</td>\n",
       "      <td>704</td>\n",
       "      <td>15084</td>\n",
       "      <td>36574</td>\n",
       "      <td>89612</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SISOG</th>\n",
       "      <td>3675</td>\n",
       "      <td>51655</td>\n",
       "      <td>90223</td>\n",
       "      <td>489</td>\n",
       "      <td>9536</td>\n",
       "      <td>23037</td>\n",
       "      <td>56162</td>\n",
       "      <td>1899593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">MM</th>\n",
       "      <th>MISOG</th>\n",
       "      <td>8933</td>\n",
       "      <td>66287</td>\n",
       "      <td>130727</td>\n",
       "      <td>616</td>\n",
       "      <td>11958</td>\n",
       "      <td>27996</td>\n",
       "      <td>67630</td>\n",
       "      <td>2960898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SISOG</th>\n",
       "      <td>3221</td>\n",
       "      <td>43801</td>\n",
       "      <td>83590</td>\n",
       "      <td>566</td>\n",
       "      <td>8437</td>\n",
       "      <td>19336</td>\n",
       "      <td>45084</td>\n",
       "      <td>2058378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          count   mean     std  min    25%    50%    75%      max\n",
       "HS MISOG   9826  82044  143420  704  15084  36574  89612  2473538\n",
       "   SISOG   3675  51655   90223  489   9536  23037  56162  1899593\n",
       "MM MISOG   8933  66287  130727  616  11958  27996  67630  2960898\n",
       "   SISOG   3221  43801   83590  566   8437  19336  45084  2058378"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "describe = pd.concat(\n",
    "    [\n",
    "        hs_genes.groupby('Miso_siso')['Gene_length'].describe().round(0).astype(int),\n",
    "        genes_new.groupby('Miso_siso')['Gene_length'].describe().round(0).astype(int),\n",
    "    ],\n",
    "    axis=0\n",
    ")\n",
    "describe.index = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ignored-watts",
   "metadata": {},
   "source": [
    "## mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "announced-correspondence",
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna = genes[['Miso_siso', 'GeneID', 'Gene name', 'ensembl_gene_id', 'ensembl_transcript_id', 'chromosome_name', 'transcript_start', 'transcript_end', 'transcript_length']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "activated-smell",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>31111</td>\n",
       "      <td>3675</td>\n",
       "      <td>22589</td>\n",
       "      <td>5314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3070</td>\n",
       "      <td>3728</td>\n",
       "      <td>3145</td>\n",
       "      <td>3116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2361</td>\n",
       "      <td>2772</td>\n",
       "      <td>2337</td>\n",
       "      <td>2297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>267</td>\n",
       "      <td>303</td>\n",
       "      <td>132</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1537</td>\n",
       "      <td>1805</td>\n",
       "      <td>1652</td>\n",
       "      <td>1577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>2424</td>\n",
       "      <td>2979</td>\n",
       "      <td>2673</td>\n",
       "      <td>2578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>3925</td>\n",
       "      <td>4865</td>\n",
       "      <td>4058</td>\n",
       "      <td>3961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>34626</td>\n",
       "      <td>30609</td>\n",
       "      <td>123179</td>\n",
       "      <td>24060</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          HS             MM       \n",
       "       MISOG  SISOG   MISOG  SISOG\n",
       "count  31111   3675   22589   5314\n",
       "mean    3070   3728    3145   3116\n",
       "std     2361   2772    2337   2297\n",
       "min      267    303     132    215\n",
       "25%     1537   1805    1652   1577\n",
       "50%     2424   2979    2673   2578\n",
       "75%     3925   4865    4058   3961\n",
       "max    34626  30609  123179  24060"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_describe = pd.concat(\n",
    "    [\n",
    "        hs_mrna.groupby('Miso_siso')['Transcript length (including UTRs and CDS)'].describe().round(0).astype(int).T,\n",
    "        mrna.groupby('Miso_siso')['transcript_length'].describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "mrna_describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "mrna_describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "concerned-scanning",
   "metadata": {},
   "source": [
    "## Exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adjusted-knight",
   "metadata": {},
   "source": [
    "### Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "improving-slovenia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>183827</td>\n",
       "      <td>36173</td>\n",
       "      <td>137283</td>\n",
       "      <td>37128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>409</td>\n",
       "      <td>378</td>\n",
       "      <td>338</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>922</td>\n",
       "      <td>962</td>\n",
       "      <td>789</td>\n",
       "      <td>777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>95</td>\n",
       "      <td>97</td>\n",
       "      <td>94</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>145</td>\n",
       "      <td>137</td>\n",
       "      <td>137</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>269</td>\n",
       "      <td>203</td>\n",
       "      <td>220</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>33289</td>\n",
       "      <td>27302</td>\n",
       "      <td>122801</td>\n",
       "      <td>20715</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           HS             MM       \n",
       "        MISOG  SISOG   MISOG  SISOG\n",
       "count  183827  36173  137283  37128\n",
       "mean      409    378     338    364\n",
       "std       922    962     789    777\n",
       "min         1      2       1      2\n",
       "25%        95     97      94     99\n",
       "50%       145    137     137    144\n",
       "75%       269    203     220    229\n",
       "max     33289  27302  122801  20715"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_exons = hs_exons.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)', 'Length'])\n",
    "hs_exons = hs_exons.loc[hs_exons['Length'] > 0]\n",
    "\n",
    "exons = exons.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)', 'Length'])\n",
    "exons = exons.loc[exons['Length'] > 0]\n",
    "\n",
    "describe = pd.concat(\n",
    "    [\n",
    "        hs_exons.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "        exons.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ranging-innocent",
   "metadata": {},
   "source": [
    "### Nb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "central-norman",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>9826</td>\n",
       "      <td>3675</td>\n",
       "      <td>8933</td>\n",
       "      <td>3221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>20</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>161</td>\n",
       "      <td>107</td>\n",
       "      <td>149</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         HS          MM      \n",
       "      MISOG SISOG MISOG SISOG\n",
       "count  9826  3675  8933  3221\n",
       "mean     16    10    14    11\n",
       "std      11    10    10    10\n",
       "min       2     2     1     1\n",
       "25%       9     4     7     5\n",
       "50%      13     7    12     8\n",
       "75%      20    12    18    14\n",
       "max     161   107   149   109"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_exons = hs_exons.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)', 'Length'])\n",
    "hs_exons = hs_exons.loc[hs_exons['Length'] > 0]\n",
    "\n",
    "exons = exons.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)', 'Length'])\n",
    "exons = exons.loc[exons['Length'] > 0]\n",
    "\n",
    "describe = pd.concat(\n",
    "    [\n",
    "        hs_exons.groupby(['Miso_siso', 'GeneID'])['Exon region start (bp)'].nunique().groupby('Miso_siso').describe().round(0).astype(int).T,\n",
    "        exons.groupby(['Miso_siso', 'GeneID'])['Exon region start (bp)'].nunique().groupby('Miso_siso').describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecological-asian",
   "metadata": {},
   "source": [
    "## CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "labeled-regression",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>161242</td>\n",
       "      <td>34776</td>\n",
       "      <td>125843</td>\n",
       "      <td>35075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>167</td>\n",
       "      <td>179</td>\n",
       "      <td>162</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>273</td>\n",
       "      <td>315</td>\n",
       "      <td>245</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>77</td>\n",
       "      <td>89</td>\n",
       "      <td>81</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>117</td>\n",
       "      <td>127</td>\n",
       "      <td>119</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>168</td>\n",
       "      <td>174</td>\n",
       "      <td>167</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>17330</td>\n",
       "      <td>15792</td>\n",
       "      <td>16625</td>\n",
       "      <td>9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           HS             MM       \n",
       "        MISOG  SISOG   MISOG  SISOG\n",
       "count  161242  34776  125843  35075\n",
       "mean      167    179     162    180\n",
       "std       273    315     245    283\n",
       "min         1      1       1      1\n",
       "25%        77     89      81     88\n",
       "50%       117    127     119    127\n",
       "75%       168    174     167    176\n",
       "max     17330  15792   16625   9658"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_cds = hs_cds.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)', 'Length'])\n",
    "hs_cds = hs_cds.loc[hs_cds['Length'] > 0]\n",
    "\n",
    "cds = cds.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)', 'Length'])\n",
    "cds = cds.loc[cds['Length'] > 0]\n",
    "\n",
    "describe = pd.concat(\n",
    "    [\n",
    "        hs_cds.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "        cds.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "spare-cooking",
   "metadata": {},
   "source": [
    "## Introns length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "micro-medicine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>142425</td>\n",
       "      <td>32498</td>\n",
       "      <td>114998</td>\n",
       "      <td>31460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7703</td>\n",
       "      <td>5005</td>\n",
       "      <td>5872</td>\n",
       "      <td>4428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>23815</td>\n",
       "      <td>14920</td>\n",
       "      <td>22222</td>\n",
       "      <td>14027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>581</td>\n",
       "      <td>458</td>\n",
       "      <td>482</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1818</td>\n",
       "      <td>1424</td>\n",
       "      <td>1442</td>\n",
       "      <td>1275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>5433</td>\n",
       "      <td>3807</td>\n",
       "      <td>3780</td>\n",
       "      <td>3211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1160410</td>\n",
       "      <td>654925</td>\n",
       "      <td>2908815</td>\n",
       "      <td>536970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            HS               MM        \n",
       "         MISOG   SISOG    MISOG   SISOG\n",
       "count   142425   32498   114998   31460\n",
       "mean      7703    5005     5872    4428\n",
       "std      23815   14920    22222   14027\n",
       "min          1       1        1       1\n",
       "25%        581     458      482     420\n",
       "50%       1818    1424     1442    1275\n",
       "75%       5433    3807     3780    3211\n",
       "max    1160410  654925  2908815  536970"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_introns = hs_introns.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Introns'])\n",
    "hs_introns = hs_introns.loc[hs_introns['Length'] > 0]\n",
    "\n",
    "introns = introns.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Introns'])\n",
    "introns = introns.loc[introns['Length'] > 0]\n",
    "\n",
    "describe = pd.concat(\n",
    "    [\n",
    "        hs_introns.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "        introns.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "disciplinary-burns",
   "metadata": {},
   "source": [
    "## Introns length according ordinal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "broad-emerald",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Miso_siso</th>\n",
       "      <th colspan=\"5\" halign=\"left\">Miso</th>\n",
       "      <th colspan=\"5\" halign=\"left\">Siso</th>\n",
       "      <th colspan=\"5\" halign=\"left\">Miso</th>\n",
       "      <th colspan=\"5\" halign=\"left\">Siso</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ordinal_nb</th>\n",
       "      <th>1.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>5.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>5.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>5.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>5.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>43680</td>\n",
       "      <td>42115</td>\n",
       "      <td>39113</td>\n",
       "      <td>35472</td>\n",
       "      <td>31887</td>\n",
       "      <td>3675</td>\n",
       "      <td>3213</td>\n",
       "      <td>2854</td>\n",
       "      <td>2532</td>\n",
       "      <td>2244</td>\n",
       "      <td>22506</td>\n",
       "      <td>21827</td>\n",
       "      <td>20503</td>\n",
       "      <td>18884</td>\n",
       "      <td>17276</td>\n",
       "      <td>5275</td>\n",
       "      <td>4719</td>\n",
       "      <td>4232</td>\n",
       "      <td>3753</td>\n",
       "      <td>3317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>15887</td>\n",
       "      <td>11778</td>\n",
       "      <td>8735</td>\n",
       "      <td>6539</td>\n",
       "      <td>5543</td>\n",
       "      <td>13167</td>\n",
       "      <td>8010</td>\n",
       "      <td>5541</td>\n",
       "      <td>4339</td>\n",
       "      <td>3735</td>\n",
       "      <td>16394</td>\n",
       "      <td>10488</td>\n",
       "      <td>7634</td>\n",
       "      <td>5518</td>\n",
       "      <td>4659</td>\n",
       "      <td>11075</td>\n",
       "      <td>7016</td>\n",
       "      <td>4772</td>\n",
       "      <td>3712</td>\n",
       "      <td>3331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>35141</td>\n",
       "      <td>31602</td>\n",
       "      <td>25903</td>\n",
       "      <td>20442</td>\n",
       "      <td>18006</td>\n",
       "      <td>29571</td>\n",
       "      <td>22329</td>\n",
       "      <td>14515</td>\n",
       "      <td>10874</td>\n",
       "      <td>8569</td>\n",
       "      <td>40173</td>\n",
       "      <td>35478</td>\n",
       "      <td>25013</td>\n",
       "      <td>17065</td>\n",
       "      <td>16075</td>\n",
       "      <td>26212</td>\n",
       "      <td>19838</td>\n",
       "      <td>12658</td>\n",
       "      <td>10007</td>\n",
       "      <td>8451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>67</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>987</td>\n",
       "      <td>802</td>\n",
       "      <td>642</td>\n",
       "      <td>579</td>\n",
       "      <td>525</td>\n",
       "      <td>1053</td>\n",
       "      <td>659</td>\n",
       "      <td>536</td>\n",
       "      <td>464</td>\n",
       "      <td>390</td>\n",
       "      <td>1000</td>\n",
       "      <td>727</td>\n",
       "      <td>569</td>\n",
       "      <td>523</td>\n",
       "      <td>457</td>\n",
       "      <td>918</td>\n",
       "      <td>635</td>\n",
       "      <td>502</td>\n",
       "      <td>432</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3998</td>\n",
       "      <td>2561</td>\n",
       "      <td>2070</td>\n",
       "      <td>1760</td>\n",
       "      <td>1643</td>\n",
       "      <td>3755</td>\n",
       "      <td>2046</td>\n",
       "      <td>1616</td>\n",
       "      <td>1434</td>\n",
       "      <td>1368</td>\n",
       "      <td>3896</td>\n",
       "      <td>2303</td>\n",
       "      <td>1755</td>\n",
       "      <td>1548</td>\n",
       "      <td>1393</td>\n",
       "      <td>3157</td>\n",
       "      <td>1813</td>\n",
       "      <td>1478</td>\n",
       "      <td>1257</td>\n",
       "      <td>1274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>15422</td>\n",
       "      <td>8326</td>\n",
       "      <td>6174</td>\n",
       "      <td>4778</td>\n",
       "      <td>4227</td>\n",
       "      <td>12537</td>\n",
       "      <td>5777</td>\n",
       "      <td>4386</td>\n",
       "      <td>3696</td>\n",
       "      <td>3490</td>\n",
       "      <td>14785</td>\n",
       "      <td>7098</td>\n",
       "      <td>4847</td>\n",
       "      <td>3851</td>\n",
       "      <td>3422</td>\n",
       "      <td>10344</td>\n",
       "      <td>4687</td>\n",
       "      <td>3718</td>\n",
       "      <td>3113</td>\n",
       "      <td>2853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>956395</td>\n",
       "      <td>1160410</td>\n",
       "      <td>772518</td>\n",
       "      <td>1096452</td>\n",
       "      <td>1047317</td>\n",
       "      <td>654925</td>\n",
       "      <td>367620</td>\n",
       "      <td>253993</td>\n",
       "      <td>222578</td>\n",
       "      <td>130369</td>\n",
       "      <td>940551</td>\n",
       "      <td>2908815</td>\n",
       "      <td>785804</td>\n",
       "      <td>551318</td>\n",
       "      <td>551318</td>\n",
       "      <td>536970</td>\n",
       "      <td>362814</td>\n",
       "      <td>171562</td>\n",
       "      <td>224742</td>\n",
       "      <td>141021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Miso_siso     Miso                                       Siso                  \\\n",
       "Ordinal_nb     1.0      2.0     3.0      4.0      5.0     1.0     2.0     3.0   \n",
       "count        43680    42115   39113    35472    31887    3675    3213    2854   \n",
       "mean         15887    11778    8735     6539     5543   13167    8010    5541   \n",
       "std          35141    31602   25903    20442    18006   29571   22329   14515   \n",
       "min             29        0       1        0        0       1      68      67   \n",
       "25%            987      802     642      579      525    1053     659     536   \n",
       "50%           3998     2561    2070     1760     1643    3755    2046    1616   \n",
       "75%          15422     8326    6174     4778     4227   12537    5777    4386   \n",
       "max         956395  1160410  772518  1096452  1047317  654925  367620  253993   \n",
       "\n",
       "Miso_siso                     Miso                                     Siso  \\\n",
       "Ordinal_nb     4.0     5.0     1.0      2.0     3.0     4.0     5.0     1.0   \n",
       "count         2532    2244   22506    21827   20503   18884   17276    5275   \n",
       "mean          4339    3735   16394    10488    7634    5518    4659   11075   \n",
       "std          10874    8569   40173    35478   25013   17065   16075   26212   \n",
       "min              1      69       1        0       4       0       0      29   \n",
       "25%            464     390    1000      727     569     523     457     918   \n",
       "50%           1434    1368    3896     2303    1755    1548    1393    3157   \n",
       "75%           3696    3490   14785     7098    4847    3851    3422   10344   \n",
       "max         222578  130369  940551  2908815  785804  551318  551318  536970   \n",
       "\n",
       "Miso_siso                                   \n",
       "Ordinal_nb     2.0     3.0     4.0     5.0  \n",
       "count         4719    4232    3753    3317  \n",
       "mean          7016    4772    3712    3331  \n",
       "std          19838   12658   10007    8451  \n",
       "min              0       3       0      64  \n",
       "25%            635     502     432     381  \n",
       "50%           1813    1478    1257    1274  \n",
       "75%           4687    3718    3113    2853  \n",
       "max         362814  171562  224742  141021  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_introns = pd.read_parquet(\"/gstock/GeneIso/V2/Introns.parquet\")\n",
    "hs_introns.loc[hs_introns['Strand'] == 1, 'Ordinal_nb_inverted'] = hs_introns.loc[hs_introns['Strand'] == 1, 'Ordinal_nb_inverted'] + 1\n",
    "hs_introns = hs_introns.loc[hs_introns['Ordinal_nb'] <= 5] \n",
    "\n",
    "introns = pd.read_parquet(\"/gstock/GeneIso/V2/Introns_MM.parquet\")\n",
    "introns.loc[introns['Strand'] == 1, 'Ordinal_nb_inverted'] = introns.loc[introns['Strand'] == 1, 'Ordinal_nb_inverted'] + 1\n",
    "introns = introns.loc[introns['Ordinal_nb'] <= 5] \n",
    "\n",
    "pd.concat(\n",
    "    [\n",
    "        hs_introns.groupby(['Miso_siso', 'Ordinal_nb'])['Length'].describe().round(0).astype(int),\n",
    "        introns.groupby(['Miso_siso', 'Ordinal_nb'])['Length'].describe().round(0).astype(int)\n",
    "\n",
    "    ], axis=0\n",
    ").T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "macro-chinese",
   "metadata": {},
   "source": [
    "## 5' UTR length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "satellite-century",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>53935</td>\n",
       "      <td>4816</td>\n",
       "      <td>27771</td>\n",
       "      <td>6633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>156</td>\n",
       "      <td>143</td>\n",
       "      <td>154</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>221</td>\n",
       "      <td>169</td>\n",
       "      <td>192</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>51</td>\n",
       "      <td>42</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>100</td>\n",
       "      <td>92</td>\n",
       "      <td>104</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>184</td>\n",
       "      <td>178</td>\n",
       "      <td>192</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>6565</td>\n",
       "      <td>2222</td>\n",
       "      <td>8961</td>\n",
       "      <td>7606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          HS           MM      \n",
       "       MISOG SISOG  MISOG SISOG\n",
       "count  53935  4816  27771  6633\n",
       "mean     156   143    154   166\n",
       "std      221   169    192   227\n",
       "min        1     1      1     1\n",
       "25%       51    42     51    54\n",
       "50%      100    92    104   108\n",
       "75%      184   178    192   205\n",
       "max     6565  2222   8961  7606"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_five_UTR = hs_five_UTR.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)'])\n",
    "hs_five_UTR = hs_five_UTR.loc[hs_five_UTR['Length'] > 0]\n",
    "\n",
    "five_UTR = five_UTR.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)'])\n",
    "five_UTR = five_UTR.loc[five_UTR['Length'] > 0]\n",
    "\n",
    "describe = pd.concat(\n",
    "    [\n",
    "        hs_five_UTR.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "        five_UTR.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brave-trance",
   "metadata": {},
   "source": [
    "## 3' UTR length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "subjective-broadway",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">HS</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "      <th>MISOG</th>\n",
       "      <th>SISOG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>20869</td>\n",
       "      <td>5546</td>\n",
       "      <td>20869</td>\n",
       "      <td>5546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1070</td>\n",
       "      <td>1130</td>\n",
       "      <td>1070</td>\n",
       "      <td>1130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1579</td>\n",
       "      <td>1411</td>\n",
       "      <td>1579</td>\n",
       "      <td>1411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>201</td>\n",
       "      <td>202</td>\n",
       "      <td>201</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>594</td>\n",
       "      <td>592</td>\n",
       "      <td>594</td>\n",
       "      <td>592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1455</td>\n",
       "      <td>1567</td>\n",
       "      <td>1455</td>\n",
       "      <td>1567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>122582</td>\n",
       "      <td>18864</td>\n",
       "      <td>122582</td>\n",
       "      <td>18864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           HS             MM       \n",
       "        MISOG  SISOG   MISOG  SISOG\n",
       "count   20869   5546   20869   5546\n",
       "mean     1070   1130    1070   1130\n",
       "std      1579   1411    1579   1411\n",
       "min         1      1       1      1\n",
       "25%       201    202     201    202\n",
       "50%       594    592     594    592\n",
       "75%      1455   1567    1455   1567\n",
       "max    122582  18864  122582  18864"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_three_UTR = hs_three_UTR.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)'])\n",
    "hs_three_UTR = hs_three_UTR.loc[hs_three_UTR['Length'] > 0]\n",
    "\n",
    "three_UTR = three_UTR.drop_duplicates(subset=['Miso_siso', 'GeneID', 'Exon region start (bp)', 'Exon region end (bp)'])\n",
    "three_UTR = three_UTR.loc[three_UTR['Length'] > 0]\n",
    "\n",
    "describe = pd.concat(\n",
    "    [\n",
    "        three_UTR.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "        three_UTR.groupby('Miso_siso')['Length'].describe().round(0).astype(int).T,\n",
    "    ],\n",
    "    axis=1\n",
    ")\n",
    "describe.columns = pd.MultiIndex.from_product([['HS', 'MM'], ['MISOG', 'SISOG']])\n",
    "describe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "numerical-pound",
   "metadata": {},
   "source": [
    "# MISOG SISOG category comparison for HS & MM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "reflected-latvia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>Miso_siso_mm</th>\n",
       "      <th>Miso_siso_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSMUSG00000067377</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSMUSG00000031250</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Siso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSMUSG00000078919</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>ENSMUSG00000026584</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>ENSMUSG00000041406</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27895</th>\n",
       "      <td>ENSG00000284194</td>\n",
       "      <td>ENSMUSG00000091780</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27897</th>\n",
       "      <td>ENSG00000284308</td>\n",
       "      <td>ENSMUSG00000030030</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27899</th>\n",
       "      <td>ENSG00000284753</td>\n",
       "      <td>ENSMUSG00000115219</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Siso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27901</th>\n",
       "      <td>ENSG00000284770</td>\n",
       "      <td>ENSMUSG00000039233</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27902</th>\n",
       "      <td>ENSG00000286140</td>\n",
       "      <td>ENSMUSG00000117748</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12154 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID     ensembl_gene_id Miso_siso Miso_siso_mm  \\\n",
       "0      ENSG00000000003  ENSMUSG00000067377      Miso         Siso   \n",
       "1      ENSG00000000005  ENSMUSG00000031250      Siso         Siso   \n",
       "2      ENSG00000000419  ENSMUSG00000078919      Miso         Miso   \n",
       "4      ENSG00000000457  ENSMUSG00000026584      Miso         Miso   \n",
       "7      ENSG00000000460  ENSMUSG00000041406      Miso         Miso   \n",
       "...                ...                 ...       ...          ...   \n",
       "27895  ENSG00000284194  ENSMUSG00000091780      Miso         Miso   \n",
       "27897  ENSG00000284308  ENSMUSG00000030030      Miso         Miso   \n",
       "27899  ENSG00000284753  ENSMUSG00000115219      Siso         Miso   \n",
       "27901  ENSG00000284770  ENSMUSG00000039233      Miso         Siso   \n",
       "27902  ENSG00000286140  ENSMUSG00000117748      Miso         Siso   \n",
       "\n",
       "        Miso_siso_full  \n",
       "0      HS:Miso-MM:Siso  \n",
       "1      HS:Siso-MM:Siso  \n",
       "2      HS:Miso-MM:Miso  \n",
       "4      HS:Miso-MM:Miso  \n",
       "7      HS:Miso-MM:Miso  \n",
       "...                ...  \n",
       "27895  HS:Miso-MM:Miso  \n",
       "27897  HS:Miso-MM:Miso  \n",
       "27899  HS:Siso-MM:Miso  \n",
       "27901  HS:Miso-MM:Siso  \n",
       "27902  HS:Miso-MM:Siso  \n",
       "\n",
       "[12154 rows x 5 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count nb of transcripts per MM gene\n",
    "new_count = genes.groupby('ensembl_gene_id')['ensembl_transcript_id'].nunique().rename('transcript_count_mm').reset_index()\n",
    "new_count.loc[new_count['transcript_count_mm'] > 1, 'Miso_siso_mm'] = 'Miso'\n",
    "new_count.loc[new_count['transcript_count_mm'] == 1, 'Miso_siso_mm'] = 'Siso'\n",
    "\n",
    "# Merge with previous DF and retrieve Miso Siso HS status\n",
    "new_count = pd.merge(genes, new_count,on='ensembl_gene_id')[['GeneID', 'ensembl_gene_id', 'Miso_siso', 'Miso_siso_mm']].drop_duplicates()\n",
    "# new_count.groupby('Miso_siso')['Miso_siso_mm'].value_counts()\n",
    "new_count['Miso_siso_full'] = 'HS:' + new_count['Miso_siso'] + '-MM:' + new_count['Miso_siso_mm']\n",
    "new_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hairy-departure",
   "metadata": {},
   "source": [
    "# MM ENCODE Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "southwest-gothic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENST</th>\n",
       "      <th>TPM_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000000001</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000000028</td>\n",
       "      <td>0.874766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000000033</td>\n",
       "      <td>0.500935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUST00000000049</td>\n",
       "      <td>0.297196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUST00000000058</td>\n",
       "      <td>0.912150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141759</th>\n",
       "      <td>ENSMUST00000238186</td>\n",
       "      <td>0.289720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141762</th>\n",
       "      <td>ENSMUST00000238189</td>\n",
       "      <td>0.596262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141763</th>\n",
       "      <td>ENSMUST00000238190</td>\n",
       "      <td>0.988785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141765</th>\n",
       "      <td>ENSMUST00000238192</td>\n",
       "      <td>0.218692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141766</th>\n",
       "      <td>ENSMUST00000238193</td>\n",
       "      <td>0.607477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91896 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ENST  TPM_ratio\n",
       "0       ENSMUST00000000001   1.000000\n",
       "3       ENSMUST00000000028   0.874766\n",
       "4       ENSMUST00000000033   0.500935\n",
       "5       ENSMUST00000000049   0.297196\n",
       "6       ENSMUST00000000058   0.912150\n",
       "...                    ...        ...\n",
       "141759  ENSMUST00000238186   0.289720\n",
       "141762  ENSMUST00000238189   0.596262\n",
       "141763  ENSMUST00000238190   0.988785\n",
       "141765  ENSMUST00000238192   0.218692\n",
       "141766  ENSMUST00000238193   0.607477\n",
       "\n",
       "[91896 rows x 2 columns]"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "encode_mm = pd.read_csv('/gstock/biolo_datasets/ENCODE/ENCODE_SRS_TPM_summary.tsv.gz', compression='gzip', sep='\\t') \n",
    "\n",
    "# Ratio = % of 535 experiments of short read seq where ENST is expressed > 0.1 TPM\n",
    "## Like GTEx, cutoff = expression > 0.1 TPM in >= 20% exp\n",
    "encode_mm = encode_mm.loc[encode_mm['TPM_ratio'] >= 0.2]\n",
    "encode_mm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "responsible-track",
   "metadata": {},
   "source": [
    "## Merge "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "described-providence",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge MM genes with MM genes expressed \n",
    "\n",
    "new_genes_mm = pd.merge(\n",
    "    genes,\n",
    "    encode_mm.rename({'ENST': 'ensembl_transcript_id'}, axis=1),\n",
    "    on='ensembl_transcript_id'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "flying-california",
   "metadata": {},
   "source": [
    "## New count expressed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "native-regulation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Same as before but to retieve expressed genes\n",
    "\n",
    "new_count = new_genes_mm.groupby('ensembl_gene_id')['ensembl_transcript_id'].nunique().rename('transcript_count_mm').reset_index()\n",
    "new_count.loc[new_count['transcript_count_mm'] > 1, 'Miso_siso_mm'] = 'Miso'\n",
    "new_count.loc[new_count['transcript_count_mm'] == 1, 'Miso_siso_mm'] = 'Siso'\n",
    "new_count = pd.merge(genes, new_count,on='ensembl_gene_id')[['GeneID', 'ensembl_gene_id', 'Miso_siso', 'Miso_siso_mm']].drop_duplicates()\n",
    "# new_count.groupby('Miso_siso')['Miso_siso_mm'].value_counts()\n",
    "new_count['Miso_siso_full'] = 'HS:' + new_count['Miso_siso'] + '-MM:' + new_count['Miso_siso_mm']\n",
    "new_count_expr = new_count.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "sixth-union",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>Miso_siso_mm</th>\n",
       "      <th>Miso_siso_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSMUSG00000067377</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSMUSG00000031250</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Siso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSMUSG00000078919</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>ENSMUSG00000026584</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>ENSMUSG00000041406</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27895</th>\n",
       "      <td>ENSG00000284194</td>\n",
       "      <td>ENSMUSG00000091780</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27897</th>\n",
       "      <td>ENSG00000284308</td>\n",
       "      <td>ENSMUSG00000030030</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27899</th>\n",
       "      <td>ENSG00000284753</td>\n",
       "      <td>ENSMUSG00000115219</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Siso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27901</th>\n",
       "      <td>ENSG00000284770</td>\n",
       "      <td>ENSMUSG00000039233</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27902</th>\n",
       "      <td>ENSG00000286140</td>\n",
       "      <td>ENSMUSG00000117748</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12154 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID     ensembl_gene_id Miso_siso Miso_siso_mm  \\\n",
       "0      ENSG00000000003  ENSMUSG00000067377      Miso         Siso   \n",
       "1      ENSG00000000005  ENSMUSG00000031250      Siso         Siso   \n",
       "2      ENSG00000000419  ENSMUSG00000078919      Miso         Miso   \n",
       "4      ENSG00000000457  ENSMUSG00000026584      Miso         Miso   \n",
       "7      ENSG00000000460  ENSMUSG00000041406      Miso         Miso   \n",
       "...                ...                 ...       ...          ...   \n",
       "27895  ENSG00000284194  ENSMUSG00000091780      Miso         Miso   \n",
       "27897  ENSG00000284308  ENSMUSG00000030030      Miso         Miso   \n",
       "27899  ENSG00000284753  ENSMUSG00000115219      Siso         Miso   \n",
       "27901  ENSG00000284770  ENSMUSG00000039233      Miso         Siso   \n",
       "27902  ENSG00000286140  ENSMUSG00000117748      Miso         Siso   \n",
       "\n",
       "        Miso_siso_full  \n",
       "0      HS:Miso-MM:Siso  \n",
       "1      HS:Siso-MM:Siso  \n",
       "2      HS:Miso-MM:Miso  \n",
       "4      HS:Miso-MM:Miso  \n",
       "7      HS:Miso-MM:Miso  \n",
       "...                ...  \n",
       "27895  HS:Miso-MM:Miso  \n",
       "27897  HS:Miso-MM:Miso  \n",
       "27899  HS:Siso-MM:Miso  \n",
       "27901  HS:Miso-MM:Siso  \n",
       "27902  HS:Miso-MM:Siso  \n",
       "\n",
       "[12154 rows x 5 columns]"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Before ENCODE expr\n",
    "new_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "compatible-salvation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>Miso_siso</th>\n",
       "      <th>Miso_siso_mm</th>\n",
       "      <th>Miso_siso_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSMUSG00000067377</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSMUSG00000031250</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Siso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSMUSG00000078919</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>ENSMUSG00000026584</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>ENSMUSG00000041406</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27024</th>\n",
       "      <td>ENSG00000284194</td>\n",
       "      <td>ENSMUSG00000091780</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27026</th>\n",
       "      <td>ENSG00000284308</td>\n",
       "      <td>ENSMUSG00000030030</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Miso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27028</th>\n",
       "      <td>ENSG00000284753</td>\n",
       "      <td>ENSMUSG00000115219</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Miso</td>\n",
       "      <td>HS:Siso-MM:Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27030</th>\n",
       "      <td>ENSG00000284770</td>\n",
       "      <td>ENSMUSG00000039233</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27031</th>\n",
       "      <td>ENSG00000286140</td>\n",
       "      <td>ENSMUSG00000117748</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>HS:Miso-MM:Siso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11693 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID     ensembl_gene_id Miso_siso Miso_siso_mm  \\\n",
       "0      ENSG00000000003  ENSMUSG00000067377      Miso         Siso   \n",
       "1      ENSG00000000005  ENSMUSG00000031250      Siso         Siso   \n",
       "2      ENSG00000000419  ENSMUSG00000078919      Miso         Miso   \n",
       "4      ENSG00000000457  ENSMUSG00000026584      Miso         Miso   \n",
       "7      ENSG00000000460  ENSMUSG00000041406      Miso         Miso   \n",
       "...                ...                 ...       ...          ...   \n",
       "27024  ENSG00000284194  ENSMUSG00000091780      Miso         Miso   \n",
       "27026  ENSG00000284308  ENSMUSG00000030030      Miso         Miso   \n",
       "27028  ENSG00000284753  ENSMUSG00000115219      Siso         Miso   \n",
       "27030  ENSG00000284770  ENSMUSG00000039233      Miso         Siso   \n",
       "27031  ENSG00000286140  ENSMUSG00000117748      Miso         Siso   \n",
       "\n",
       "        Miso_siso_full  \n",
       "0      HS:Miso-MM:Siso  \n",
       "1      HS:Siso-MM:Siso  \n",
       "2      HS:Miso-MM:Miso  \n",
       "4      HS:Miso-MM:Miso  \n",
       "7      HS:Miso-MM:Miso  \n",
       "...                ...  \n",
       "27024  HS:Miso-MM:Miso  \n",
       "27026  HS:Miso-MM:Miso  \n",
       "27028  HS:Siso-MM:Miso  \n",
       "27030  HS:Miso-MM:Siso  \n",
       "27031  HS:Miso-MM:Siso  \n",
       "\n",
       "[11693 rows x 5 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# After ENCODE expr\n",
    "new_count_expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "killing-enough",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>Miso_siso_mm</th>\n",
       "      <th>Miso_siso_mm_expr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSMUSG00000067377</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSMUSG00000031250</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSMUSG00000078919</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>ENSMUSG00000026584</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>ENSMUSG00000041406</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11688</th>\n",
       "      <td>ENSG00000284194</td>\n",
       "      <td>ENSMUSG00000091780</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11689</th>\n",
       "      <td>ENSG00000284308</td>\n",
       "      <td>ENSMUSG00000030030</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11690</th>\n",
       "      <td>ENSG00000284753</td>\n",
       "      <td>ENSMUSG00000115219</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Miso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11691</th>\n",
       "      <td>ENSG00000284770</td>\n",
       "      <td>ENSMUSG00000039233</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11692</th>\n",
       "      <td>ENSG00000286140</td>\n",
       "      <td>ENSMUSG00000117748</td>\n",
       "      <td>Siso</td>\n",
       "      <td>Siso</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11693 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID     ensembl_gene_id Miso_siso_mm Miso_siso_mm_expr\n",
       "0      ENSG00000000003  ENSMUSG00000067377         Siso              Siso\n",
       "1      ENSG00000000005  ENSMUSG00000031250         Siso              Siso\n",
       "2      ENSG00000000419  ENSMUSG00000078919         Miso              Miso\n",
       "3      ENSG00000000457  ENSMUSG00000026584         Miso              Miso\n",
       "4      ENSG00000000460  ENSMUSG00000041406         Miso              Miso\n",
       "...                ...                 ...          ...               ...\n",
       "11688  ENSG00000284194  ENSMUSG00000091780         Miso              Miso\n",
       "11689  ENSG00000284308  ENSMUSG00000030030         Miso              Miso\n",
       "11690  ENSG00000284753  ENSMUSG00000115219         Miso              Miso\n",
       "11691  ENSG00000284770  ENSMUSG00000039233         Siso              Siso\n",
       "11692  ENSG00000286140  ENSMUSG00000117748         Siso              Siso\n",
       "\n",
       "[11693 rows x 4 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge to compare MM genes that change of status\n",
    "\n",
    "merge_merge = pd.merge(\n",
    "    new_count[['GeneID', 'ensembl_gene_id', 'Miso_siso_mm']],\n",
    "    new_count_expr[['GeneID', 'ensembl_gene_id', 'Miso_siso_mm']].rename({'Miso_siso_mm' : 'Miso_siso_mm_expr'}, axis=1),\n",
    "    on=['GeneID', 'ensembl_gene_id']\n",
    ")\n",
    "\n",
    "# Export to reload later\n",
    "merge_merge.to_csv('/gstock/GeneIso/V2_with_MM_expression/misog_sisog_mm_expr.tsv.gz', compression='gzip', sep='\\t', index=False)\n",
    "merge_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "passing-render",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>Miso_siso_mm</th>\n",
       "      <th>Miso_siso_mm_expr</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004848</td>\n",
       "      <td>ENSMUSG00000035277</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>ARX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000005381</td>\n",
       "      <td>ENSMUSG00000009350</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>MPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000006128</td>\n",
       "      <td>ENSMUSG00000061762</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>TAC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000006611</td>\n",
       "      <td>ENSMUSG00000030838</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>USH1C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000007047</td>\n",
       "      <td>ENSMUSG00000030397</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>MARK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1472</th>\n",
       "      <td>ENSG00000268089</td>\n",
       "      <td>ENSMUSG00000031344</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>GABRQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1474</th>\n",
       "      <td>ENSG00000272514</td>\n",
       "      <td>ENSMUSG00000028294</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>CFAP206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>ENSG00000273079</td>\n",
       "      <td>ENSMUSG00000030209</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>GRIN2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>ENSG00000282608</td>\n",
       "      <td>ENSMUSG00000000562</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>ADORA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1481</th>\n",
       "      <td>ENSG00000283297</td>\n",
       "      <td>ENSMUSG00000079304</td>\n",
       "      <td>Miso</td>\n",
       "      <td>Siso</td>\n",
       "      <td>TEX52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>588 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID     ensembl_gene_id Miso_siso_mm Miso_siso_mm_expr  \\\n",
       "0     ENSG00000004848  ENSMUSG00000035277         Miso              Siso   \n",
       "2     ENSG00000005381  ENSMUSG00000009350         Miso              Siso   \n",
       "4     ENSG00000006128  ENSMUSG00000061762         Miso              Siso   \n",
       "7     ENSG00000006611  ENSMUSG00000030838         Miso              Siso   \n",
       "12    ENSG00000007047  ENSMUSG00000030397         Miso              Siso   \n",
       "...               ...                 ...          ...               ...   \n",
       "1472  ENSG00000268089  ENSMUSG00000031344         Miso              Siso   \n",
       "1474  ENSG00000272514  ENSMUSG00000028294         Miso              Siso   \n",
       "1476  ENSG00000273079  ENSMUSG00000030209         Miso              Siso   \n",
       "1479  ENSG00000282608  ENSMUSG00000000562         Miso              Siso   \n",
       "1481  ENSG00000283297  ENSMUSG00000079304         Miso              Siso   \n",
       "\n",
       "     Gene name  \n",
       "0          ARX  \n",
       "2          MPO  \n",
       "4         TAC1  \n",
       "7        USH1C  \n",
       "12       MARK4  \n",
       "...        ...  \n",
       "1472     GABRQ  \n",
       "1474   CFAP206  \n",
       "1476    GRIN2B  \n",
       "1479    ADORA3  \n",
       "1481     TEX52  \n",
       "\n",
       "[588 rows x 5 columns]"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify MM genes that change of status before and after ENCODE expr\n",
    "changing_status_mm_genes = pd.merge(\n",
    "    merge_merge.loc[(merge_merge['Miso_siso_mm'] == \"Miso\") & (merge_merge['Miso_siso_mm_expr'] == \"Siso\")],\n",
    "    genes[['GeneID', 'Gene name']],\n",
    "    on='GeneID'\n",
    ").drop_duplicates()\n",
    "changing_status_mm_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mechanical-deposit",
   "metadata": {},
   "source": [
    "## Count of genes that change after expression in MM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "essential-fairy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso_siso_mm  Miso_siso_mm_expr\n",
       "Miso          Miso                 6608\n",
       "              Siso                  588\n",
       "Siso          Siso                 4497\n",
       "Name: Miso_siso_mm_expr, dtype: int64"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MM genes before & after expr\n",
    "merge_merge.groupby(['Miso_siso_mm'])['Miso_siso_mm_expr'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "external-bikini",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Miso_siso  Miso_siso_mm\n",
       "Miso       Miso            5490\n",
       "           Siso            3213\n",
       "Siso       Siso            1872\n",
       "           Miso            1118\n",
       "Name: Miso_siso_mm, dtype: int64"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Final distribution for Miso Siso HS & Miso Siso MM after expr\n",
    "new_count.groupby('Miso_siso')['Miso_siso_mm'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "aging-needle",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HS:Miso-MM:Siso\n",
      "/gstock/GeneIso/V2_with_MM_expression/Gene_lists_STRING/HS-Miso-MM-Siso_HS_gene_names.txt\n",
      "HS:Siso-MM:Siso\n",
      "/gstock/GeneIso/V2_with_MM_expression/Gene_lists_STRING/HS-Siso-MM-Siso_HS_gene_names.txt\n",
      "HS:Miso-MM:Miso\n",
      "/gstock/GeneIso/V2_with_MM_expression/Gene_lists_STRING/HS-Miso-MM-Miso_HS_gene_names.txt\n",
      "HS:Siso-MM:Miso\n",
      "/gstock/GeneIso/V2_with_MM_expression/Gene_lists_STRING/HS-Siso-MM-Miso_HS_gene_names.txt\n"
     ]
    }
   ],
   "source": [
    "# Prepare for export to process with stringdb command of laetitia\n",
    "\n",
    "new_count_to_export = pd.merge(new_count, genes[['GeneID', 'Gene name', 'mmusculus_homolog_associated_gene_name']].drop_duplicates(), on='GeneID')\n",
    "\n",
    "for condition in new_count_to_export.Miso_siso_full.unique().tolist():\n",
    "    print(condition)\n",
    "    print('/gstock/GeneIso/V2_with_MM_expression/Gene_lists_STRING/{}_HS_gene_names.txt'.format(condition.replace(':', '-')))\n",
    "    new_count_to_export.loc[new_count_to_export['Miso_siso_full'] == condition, 'Gene name'].to_csv('/gstock/GeneIso/V2_with_MM_expression/Gene_lists_STRING/{}_HS_gene_names.txt'.format(condition.replace(':', '-')), index=False, header=False)\n",
    "#     new_count_to_export.loc[new_count_to_export['Miso_siso_full'] == condition, 'mmusculus_homolog_associated_gene_name'].to_csv('/gstock/GeneIso/V2/Gene_lists_STRING/{}_MM_gene_names.txt'.format(condition.replace(':', '-')), index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "6bc58787a7e3b85efa746fe1d312329650c66964b196f4af9721c9b0ded7fae1"
  },
  "kernelspec": {
   "display_name": "Python [conda env:.conda-ExoCarto] *",
   "language": "python",
   "name": "conda-env-.conda-ExoCarto-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
